{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7bc4bad6-861f-45cd-b532-e33c7ec59c42",
   "metadata": {},
   "source": [
    "# Advanced Querying Techniques in Elasticsearch for BiorXiv Data Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca714fcb-2565-4dc6-8ad4-b6dac4bec88a",
   "metadata": {},
   "source": [
    "Elasticsearch, renowned for its speed and scalability, is an indispensable tool for data scientists and researchers working with large datasets like those from BiorXiv. BiorXiv provides a rich corpus of preprint publications in the life sciences, offering a wealth of data for analysis. In this guide, we delve into sophisticated querying techniques to extract meaningful insights from BiorXiv data using Elasticsearch. We'll explore everything from basic keyword searches to complex aggregations and filters."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57942729-500c-4e45-8d13-ef610eaf1dbf",
   "metadata": {},
   "source": [
    "### Establishing a Secure Connection to Elasticsearch"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5567d6f-2f83-4c25-99e0-0846966dc69b",
   "metadata": {},
   "source": [
    "\n",
    "Before diving into queries, establishing a secure connection to your Elasticsearch cluster is paramount. This ensures that your data interactions are encrypted and protected. Here’s a brief refresher on setting up a secure connection:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a76637ec-6312-45e7-aee5-e89be9a4b18e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 5.07 ms, sys: 199 µs, total: 5.27 ms\n",
      "Wall time: 4.04 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "import ssl\n",
    "\n",
    "# Path to the CA certificate\n",
    "ca_cert_path = '/workspace/repos/osl/rxiv-restapi/containers/esconfig/certs/http_ca.crt'\n",
    "# Create an SSL context\n",
    "ssl_context = ssl.create_default_context(cafile=ca_cert_path)\n",
    "# Create a connection to Elasticsearch with authentication and SSL context\n",
    "es = Elasticsearch(\n",
    "    [\"https://es:9200\"],\n",
    "    basic_auth=(\"elastic\", \"worksfine\"),\n",
    "    ssl_context=ssl_context\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3e9203a-5988-409b-ae62-8c1c8d3f6e6a",
   "metadata": {},
   "source": [
    "### Index JSON Data into Elasticsearch\n",
    "**Before we can search, you must have the data indexed in Elasticsearch.** \n",
    "**Here's a simplified function to read the JSON file and index its contents.** \n",
    "**This example assumes that your JSON data is an array of objects, each representing a document to be indexed.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "5e6d3258-5682-41e7-b3b6-c406bb9bd8d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "from elasticsearch import Elasticsearch\n",
    "\n",
    "def index_json_data(es: Elasticsearch, file_path: str, index_name: str) -> None:\n",
    "    \"\"\"\n",
    "    Reads data from a JSON file and indexes it into Elasticsearch.\n",
    "\n",
    "    Parameters:\n",
    "    es (Elasticsearch): An Elasticsearch client instance.\n",
    "    file_path (str): The path to the JSON file.\n",
    "    index_name (str): The name of the Elasticsearch index where data will be stored.\n",
    "    \"\"\"\n",
    "    # Load JSON data from the file\n",
    "    with open(file_path, 'r', encoding='utf-8') as file:\n",
    "        data = json.load(file)\n",
    "\n",
    "    # Assuming `data` is a list of documents\n",
    "    for doc in data:\n",
    "        # Index each document\n",
    "        res = es.index(index=index_name, document=doc)\n",
    "        # print(res['result'])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46318043-4de4-4dc2-9edd-21f93c56e11f",
   "metadata": {},
   "source": [
    "### Indexing Data into Elasticsearch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "bec3e056-e980-44b1-9533-eda980b86747",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 3min 28s, sys: 18 s, total: 3min 46s\n",
      "Wall time: 27min 26s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "# Assuming index_json_data is a previously defined function that indexes data from a JSON file to Elasticsearch\n",
    "# Path to the JSON file containing data to be indexed\n",
    "file_path = '/workspace/repos/osl/rxiv-restapi/docs/notebooks/data/biorxiv_2022-01-01_2024-01-11.json'\n",
    "# Name of the Elasticsearch index\n",
    "index_name = 'biorxiv'\n",
    "# Index data from the specified JSON file into Elasticsearch\n",
    "index_json_data(es, file_path, index_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ca8104d-2a52-4e2b-b7b5-36961a09b2f5",
   "metadata": {},
   "source": [
    "## Querying Analyses for BiorXiv Data\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "438f5163-489c-4ea3-92db-a8e82889ebdb",
   "metadata": {},
   "source": [
    "### Create a Search Function\n",
    "\n",
    "**After indexing the data, we can create a function to perform searches using the Elasticsearch client.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "1cf28c07-5567-4cb6-b2e8-1f540cf23f1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def search_data(es: Elasticsearch, index_name: str, query: dict) -> list:\n",
    "    \"\"\"\n",
    "    Performs a search query in an Elasticsearch index and returns only the document data.\n",
    "    search_data function is structured to perform a search query and return document hits directly,\n",
    "\n",
    "    Parameters:\n",
    "    es (Elasticsearch): An Elasticsearch client instance.\n",
    "    index_name (str): The name of the Elasticsearch index to search in.\n",
    "    query (dict): The search query in Elasticsearch Query DSL format.\n",
    "\n",
    "    Returns:\n",
    "    list: A list of documents from the search results, each represented as a dictionary.\n",
    "    \"\"\"\n",
    "    try:\n",
    "        response = es.search(index=index_name, query=query)\n",
    "        # Extract only the '_source' from each hit\n",
    "        documents = [hit['_source'] for hit in response['hits']['hits']]\n",
    "        return documents\n",
    "    except Exception as e:\n",
    "        print(f\"Search failed: {e}\")\n",
    "        return []"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8fddfd1d-9bc8-4962-b268-c196aa8b1805",
   "metadata": {},
   "source": [
    "### Keyword Searches: The Basics\n",
    "#### Search for documents with a specific title.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c79806b0-e503-4bfa-95c3-6ea93d397720",
   "metadata": {},
   "source": [
    "These query examples illustrate the flexibility of Elasticsearch's Query DSL to retrieve specific data based on various search criteria.\n",
    "\n",
    "Creating a variety of Elasticsearch query combinations involves using different aspects of the Elasticsearch Query DSL (Domain Specific Language) to retrieve specific documents based on your criteria. Below are several examples of query combinations that can be used to retrieve keys and values from the JSON data provided\n",
    "\n",
    "Each of these queries can be passed to the `search_data` function you've defined to retrieve documents from Elasticsearch based on the specified criteria. Remember to replace `index_name` with the name of your index when calling the function:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea001516-8b65-4ac9-aac6-92631c0538b7",
   "metadata": {},
   "source": [
    "### Match Query\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58875093-badd-4145-a82a-e7e80b7973ce",
   "metadata": {},
   "source": [
    "Keyword searches are the foundation of data retrieval in Elasticsearch. For instance, finding all BiorXiv papers related to CRISPR:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "8b132847-794d-49e7-bf68-492372dbed01",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Results:  10 \n",
      "\n",
      "{'doi': '10.1101/2024.01.05.574328', 'title': 'CRISPR-repressed toxin-antitoxin provides population-level immunity against diverse anti-CRISPR elements', 'authors': 'Li, M.; Shu, X.; Wang, R.; Li, Z.; Xue, Q.; Liu, C.; Cheng, F.; Zhao, H.; Wang, J.; Liu, J.; Hu, C.; Li, J.; Ouyang, S.', 'author_corresponding': 'Ming Li', 'author_corresponding_institution': 'Institute of Microbiology, CAS', 'date': '2024-01-05', 'version': '1', 'license': 'cc_no', 'category': 'Microbiology', 'jatsxml': 'https://www.biorxiv.org/content/early/2024/01/05/2024.01.05.574328.source.xml', 'abstract': 'Prokaryotic CRISPR-Cas systems are highly vulnerable to phage-encoded anti-CRISPR (Acr) factors. How CRISPR-Cas systems protect themselves remains unclear. Here, we uncovered a broad-spectrum anti-anti-CRISPR strategy involving a phage-derived toxic protein. Transcription of this toxin is normally reppressed by the CRISPR-Cas effector, but is activated to halt cell division when the effector is inhibited by any anti-CRISPR proteins or RNAs. We showed that this abortive infection-like effect efficiently expels Acr elements from bacterial population. Furthermore, we exploited this anti-anti-CRISPR mechanism to develop a screening method for specific Acr candidates for a CRISPR-Cas system, and successfully identified two distinct Acr proteins that enhance the binding of CRISPR effector to non-target DNA. Our data highlight the broad-spectrum role of CRISPR-repressed toxins in counteracting various types of Acr factors, which illuminates that the regulatory function of CRISPR-Cas confers host cells herd immunity against Acr-encoding genetic invaders, no matter they are CRISPR-targeted or not.', 'node': 103416, 'link_page': 'https://www.biorxiv.org/content/10.1101/2024.01.05.574328v1?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2024.01.05.574328v1.full.pdf'}\n",
      "{'doi': '10.1101/2024.01.05.574328', 'title': 'CRISPR-repressed toxin-antitoxin provides population-level immunity against diverse anti-CRISPR elements', 'authors': 'Li, M.; Shu, X.; Wang, R.; Li, Z.; Xue, Q.; Liu, C.; Cheng, F.; Zhao, H.; Wang, J.; Liu, J.; Hu, C.; Li, J.; Ouyang, S.', 'author_corresponding': 'Ming Li', 'author_corresponding_institution': 'Institute of Microbiology, CAS', 'date': '2024-01-05', 'version': '1', 'license': 'cc_no', 'category': 'Microbiology', 'jatsxml': 'https://www.biorxiv.org/content/early/2024/01/05/2024.01.05.574328.source.xml', 'abstract': 'Prokaryotic CRISPR-Cas systems are highly vulnerable to phage-encoded anti-CRISPR (Acr) factors. How CRISPR-Cas systems protect themselves remains unclear. Here, we uncovered a broad-spectrum anti-anti-CRISPR strategy involving a phage-derived toxic protein. Transcription of this toxin is normally reppressed by the CRISPR-Cas effector, but is activated to halt cell division when the effector is inhibited by any anti-CRISPR proteins or RNAs. We showed that this abortive infection-like effect efficiently expels Acr elements from bacterial population. Furthermore, we exploited this anti-anti-CRISPR mechanism to develop a screening method for specific Acr candidates for a CRISPR-Cas system, and successfully identified two distinct Acr proteins that enhance the binding of CRISPR effector to non-target DNA. Our data highlight the broad-spectrum role of CRISPR-repressed toxins in counteracting various types of Acr factors, which illuminates that the regulatory function of CRISPR-Cas confers host cells herd immunity against Acr-encoding genetic invaders, no matter they are CRISPR-targeted or not.', 'node': 103416, 'link_page': 'https://www.biorxiv.org/content/10.1101/2024.01.05.574328v1?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2024.01.05.574328v1.full.pdf'}\n",
      "{'doi': '10.1101/2023.03.24.534125', 'title': 'Cbp1-Cren7 chromatinization of CRISPR arrays favours transcription from leader- over cryptic promoters', 'authors': 'Blombach, F.; Sykora, M.; Case, J.; Feng, X.; Baquero, D. P.; Fouqueau, T.; Phung, D. K.; Barker, D.; Krupovic, M.; She, Q.; Werner, F.', 'author_corresponding': 'Finn Werner', 'author_corresponding_institution': 'University College London', 'date': '2023-03-25', 'version': '1', 'license': 'cc_by_nc_nd', 'category': 'Microbiology', 'jatsxml': 'https://www.biorxiv.org/content/early/2023/03/25/2023.03.24.534125.source.xml', 'abstract': 'CRISPR arrays form the physical memory of CRISPR adaptive immune systems by incorporating foreign DNA as spacers that are often AT-rich and derived from viruses. As promoter elements such as the TATA-box are AT-rich, CRISPR arrays are prone to harbouring cryptic promoters. Sulfolobales harbor extremely long CRISPR arrays spanning several kilobases, a feature that is accompanied by the CRISPR-specific transcription factor Cbp1. Aberrant Cbp1 expression modulates CRISPR array transcription, but the molecular mechanisms underlying this regulation are unknown. Here, we characterise the genome-wide Cbp1 binding at nucleotide resolution and characterise the binding motifs on distinct CRISPR arrays, as well as on unexpected non-canonical binding sites associated with transposasons. Cbp1 recruits Cren7 forming together  chimeric chromatin-like structures at CRISPR arrays. We dissect Cbp1 function in vitro and in vivo and show that the third HTH domain is responsible for Cren7 recruitment, and that Cbp1-Cren7 chromatinization plays a dual role in the transcription of CRISPR arrays. It suppresses spurious transcription from cryptic promoters within CRISPR arrays but enhances CRISPR RNA transcription directed from their cognate promoters in their leader region. Our results show that Cbp1-Cren7 chromatinization drives the productive expression of long CRISPR arrays.', 'published': 'NA', 'node': 66181, 'link_page': 'https://www.biorxiv.org/content/10.1101/2023.03.24.534125v1?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2023.03.24.534125v1.full.pdf'}\n",
      "{'doi': '10.1101/2023.03.24.534125', 'title': 'Cbp1-Cren7 chromatinization of CRISPR arrays favours transcription from leader- over cryptic promoters', 'authors': 'Blombach, F.; Sykora, M.; Case, J.; Feng, X.; Baquero, D. P.; Fouqueau, T.; Phung, D. K.; Barker, D.; Krupovic, M.; She, Q.; Werner, F.', 'author_corresponding': 'Finn Werner', 'author_corresponding_institution': 'University College London', 'date': '2023-03-25', 'version': '1', 'license': 'cc_by_nc_nd', 'category': 'Microbiology', 'jatsxml': 'https://www.biorxiv.org/content/early/2023/03/25/2023.03.24.534125.source.xml', 'abstract': 'CRISPR arrays form the physical memory of CRISPR adaptive immune systems by incorporating foreign DNA as spacers that are often AT-rich and derived from viruses. As promoter elements such as the TATA-box are AT-rich, CRISPR arrays are prone to harbouring cryptic promoters. Sulfolobales harbor extremely long CRISPR arrays spanning several kilobases, a feature that is accompanied by the CRISPR-specific transcription factor Cbp1. Aberrant Cbp1 expression modulates CRISPR array transcription, but the molecular mechanisms underlying this regulation are unknown. Here, we characterise the genome-wide Cbp1 binding at nucleotide resolution and characterise the binding motifs on distinct CRISPR arrays, as well as on unexpected non-canonical binding sites associated with transposasons. Cbp1 recruits Cren7 forming together  chimeric chromatin-like structures at CRISPR arrays. We dissect Cbp1 function in vitro and in vivo and show that the third HTH domain is responsible for Cren7 recruitment, and that Cbp1-Cren7 chromatinization plays a dual role in the transcription of CRISPR arrays. It suppresses spurious transcription from cryptic promoters within CRISPR arrays but enhances CRISPR RNA transcription directed from their cognate promoters in their leader region. Our results show that Cbp1-Cren7 chromatinization drives the productive expression of long CRISPR arrays.', 'published': 'NA', 'node': 66181, 'link_page': 'https://www.biorxiv.org/content/10.1101/2023.03.24.534125v1?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2023.03.24.534125v1.full.pdf'}\n",
      "{'doi': '10.1101/2023.03.19.533347', 'title': 'Endogenous Staphylococcus aureus CRISPR-cas system limits phage proliferation and efficiently excises from the genome as part of the SCCmec cassette', 'authors': 'Mikkelsen, K.; Bowring, J. Z.; Ng, Y. K.; Frisinger, F. S.; Maglegaard, J. K.; Li, Q.; Sieber, R. N.; Petersen, A.; Andersen, P. S.; Rostoel, J. T.; Hoeyland-Kroghsbo, N. M.; Ingmer, H.', 'author_corresponding': 'Hanne Ingmer', 'author_corresponding_institution': 'University  of Copenhagen', 'date': '2023-03-19', 'version': '1', 'license': 'cc_no', 'category': 'Microbiology', 'jatsxml': 'https://www.biorxiv.org/content/early/2023/03/19/2023.03.19.533347.source.xml', 'abstract': 'CRISPR-Cas is an adaptive immune system that allows bacteria to inactivate mobile genetic elements. Approximately 50% of bacteria harbor CRISPR-cas, however in the human pathogen Staphylococcus aureus, CRISPR-cas loci are less common and often studied in heterologous systems. We analyzed the prevalence of CRISPR-cas in genomes of methicillin resistant Staphylococcus aureus (MRSA) isolated in Denmark. Only 2.9 % of the strains carried CRISPR-cas systems, but for strains of sequence type ST630 over half were positive. All CRISPR-cas loci were type III-A and located within the staphylococcal chromosomal cassette (SCCmec) type V(5C2&5) conferring {beta}-lactam resistance. Curiously, only 23 different CRISPR spacers were identified in 69 CRISPR-positive strains and almost identical SCCmec cassettes, CRISPR arrays and cas genes, are present in staphylococcal species other than aureus, suggesting that these were transferred horizontally. For the ST630 strain 110900, we demonstrate that the SCCmec cassette containing CRISPR-cas excises from the chromosome at high frequency. However, the cassette was not transferable under the conditions investigated. One of the CRISPR spacers targets a late gene in the lytic bacteriophage (phage) virus philPLA-RODI, and we show that the system protects against phage infection by reducing phage burst size. However, CRISPR-Cas can be overloaded or bypassed by CRISPR escape mutants. Our results imply that the endogenous type III-A CRISPR-Cas system in S. aureus is active against targeted phages, albeit with low efficacy. This suggests native S. aureus CRISPR-Cas offers only partial immunity, and in nature may work in tandem with other defense systems.\\n\\nImportanceCRISPR-Cas is an adaptive immune system enabling bacteria and archaea to protect themselves against mobile genetic elements such as phages. In strains of Staphylococcus aureus, CRISPR-cas is rare, but when present, it is located within the SCCmec element encoding resistance to methicillin and other {beta}-lactam antibiotics. We show that the entire module is excisable, with almost identical versions found in different species of non-aureus staphylococci suggesting that the system only rarely acquires new spacers in S. aureus. Additionally, we show that in its endogenous form, the S. aureus CRISPR-Cas is active but inefficient against lytic phages, with phages being able to form escape mutants or overload the system. This leads us to propose that CRISPR-Cas in S. aureus offers only partial immunity in native systems, and so may work together with other defense systems to prevent phage-mediated killing.', 'published': 'NA', 'node': 65190, 'link_page': 'https://www.biorxiv.org/content/10.1101/2023.03.19.533347v1?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2023.03.19.533347v1.full.pdf'}\n",
      "{'doi': '10.1101/2023.03.19.533347', 'title': 'Endogenous Staphylococcus aureus CRISPR-cas system limits phage proliferation and efficiently excises from the genome as part of the SCCmec cassette', 'authors': 'Mikkelsen, K.; Bowring, J. Z.; Ng, Y. K.; Frisinger, F. S.; Maglegaard, J. K.; Li, Q.; Sieber, R. N.; Petersen, A.; Andersen, P. S.; Rostoel, J. T.; Hoeyland-Kroghsbo, N. M.; Ingmer, H.', 'author_corresponding': 'Hanne Ingmer', 'author_corresponding_institution': 'University  of Copenhagen', 'date': '2023-03-19', 'version': '1', 'license': 'cc_no', 'category': 'Microbiology', 'jatsxml': 'https://www.biorxiv.org/content/early/2023/03/19/2023.03.19.533347.source.xml', 'abstract': 'CRISPR-Cas is an adaptive immune system that allows bacteria to inactivate mobile genetic elements. Approximately 50% of bacteria harbor CRISPR-cas, however in the human pathogen Staphylococcus aureus, CRISPR-cas loci are less common and often studied in heterologous systems. We analyzed the prevalence of CRISPR-cas in genomes of methicillin resistant Staphylococcus aureus (MRSA) isolated in Denmark. Only 2.9 % of the strains carried CRISPR-cas systems, but for strains of sequence type ST630 over half were positive. All CRISPR-cas loci were type III-A and located within the staphylococcal chromosomal cassette (SCCmec) type V(5C2&5) conferring {beta}-lactam resistance. Curiously, only 23 different CRISPR spacers were identified in 69 CRISPR-positive strains and almost identical SCCmec cassettes, CRISPR arrays and cas genes, are present in staphylococcal species other than aureus, suggesting that these were transferred horizontally. For the ST630 strain 110900, we demonstrate that the SCCmec cassette containing CRISPR-cas excises from the chromosome at high frequency. However, the cassette was not transferable under the conditions investigated. One of the CRISPR spacers targets a late gene in the lytic bacteriophage (phage) virus philPLA-RODI, and we show that the system protects against phage infection by reducing phage burst size. However, CRISPR-Cas can be overloaded or bypassed by CRISPR escape mutants. Our results imply that the endogenous type III-A CRISPR-Cas system in S. aureus is active against targeted phages, albeit with low efficacy. This suggests native S. aureus CRISPR-Cas offers only partial immunity, and in nature may work in tandem with other defense systems.\\n\\nImportanceCRISPR-Cas is an adaptive immune system enabling bacteria and archaea to protect themselves against mobile genetic elements such as phages. In strains of Staphylococcus aureus, CRISPR-cas is rare, but when present, it is located within the SCCmec element encoding resistance to methicillin and other {beta}-lactam antibiotics. We show that the entire module is excisable, with almost identical versions found in different species of non-aureus staphylococci suggesting that the system only rarely acquires new spacers in S. aureus. Additionally, we show that in its endogenous form, the S. aureus CRISPR-Cas is active but inefficient against lytic phages, with phages being able to form escape mutants or overload the system. This leads us to propose that CRISPR-Cas in S. aureus offers only partial immunity in native systems, and so may work together with other defense systems to prevent phage-mediated killing.', 'published': 'NA', 'node': 65190, 'link_page': 'https://www.biorxiv.org/content/10.1101/2023.03.19.533347v1?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2023.03.19.533347v1.full.pdf'}\n",
      "{'doi': '10.1101/2022.04.21.488824', 'title': 'A comprehensive Bioconductor ecosystem for the design of CRISPR guide RNAs across nucleases and technologies', 'authors': 'Hoberecht, L.; Perampalam, P.; Lun, A.; Fortin, J.-P.', 'author_corresponding': 'Jean-Philippe  Fortin', 'author_corresponding_institution': 'Genentech', 'date': '2022-04-22', 'version': '1', 'license': 'cc_by_nc_nd', 'category': 'Bioinformatics', 'jatsxml': 'https://www.biorxiv.org/content/early/2022/04/22/2022.04.21.488824.source.xml', 'abstract': 'The success of CRISPR-mediated gene perturbation studies is highly dependent on the quality of gRNAs, and several tools have been developed to enable optimal gRNA design. However, these tools are not all adaptable to the latest CRISPR modalities or nucleases, nor do they offer comprehensive annotation methods for advanced CRISPR applications. Here, we present a new ecosystem of R packages, called crispr-Verse, that enables efficient gRNA design and annotation for a multitude of CRISPR technologies. This includes CRISPR knockout (CRISPRko), CRISPR activation (CRISPRa), CRISPR interference (CRISPRi), CRISPR base editing (CRISPRbe) and CRISPR knockdown (CRISPRkd). The core package, crisprDesign, offers a comprehensive, user-friendly, and unified interface to add on- and off-target annotations via several alignment methods, rich gene and SNP annotations, and a dozen on- and off-target activity scores. These functionalities are enabled for any RNA- or DNA-targeting nucleases, including Cas9, Cas12, and Cas13. We illustrate the general applicability of our tools by designing optimal gRNAs for three case studies: tiling CRISPRbe library for BRCA1 using the base editor BE4max, tiling RNA-targeting libraries for CD46 and CD55 using CasRx, and activation of MMP7 using CRISPRa. The crisprVerse ecosystem is open-source and deployed through the Bioconductor project to facilitate its use by the CRISPR community (https://github.com/crisprVerse).', 'published': '10.1038/s41467-022-34320-7', 'node': 20808, 'link_page': 'https://www.biorxiv.org/content/10.1101/2022.04.21.488824v1?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2022.04.21.488824v1.full.pdf'}\n",
      "{'doi': '10.1101/2022.04.21.488824', 'title': 'A comprehensive Bioconductor ecosystem for the design of CRISPR guide RNAs across nucleases and technologies', 'authors': 'Hoberecht, L.; Perampalam, P.; Lun, A.; Fortin, J.-P.', 'author_corresponding': 'Jean-Philippe  Fortin', 'author_corresponding_institution': 'Genentech', 'date': '2022-04-28', 'version': '2', 'license': 'cc_by_nc_nd', 'category': 'Bioinformatics', 'jatsxml': 'https://www.biorxiv.org/content/early/2022/04/28/2022.04.21.488824.source.xml', 'abstract': 'The success of CRISPR-mediated gene perturbation studies is highly dependent on the quality of gRNAs, and several tools have been developed to enable optimal gRNA design. However, these tools are not all adaptable to the latest CRISPR modalities or nucleases, nor do they offer comprehensive annotation methods for advanced CRISPR applications. Here, we present a new ecosystem of R packages, called crispr-Verse, that enables efficient gRNA design and annotation for a multitude of CRISPR technologies. This includes CRISPR knockout (CRISPRko), CRISPR activation (CRISPRa), CRISPR interference (CRISPRi), CRISPR base editing (CRISPRbe) and CRISPR knockdown (CRISPRkd). The core package, crisprDesign, offers a comprehensive, user-friendly, and unified interface to add on- and off-target annotations via several alignment methods, rich gene and SNP annotations, and a dozen on- and off-target activity scores. These functionalities are enabled for any RNA- or DNA-targeting nucleases, including Cas9, Cas12, and Cas13. We illustrate the general applicability of our tools by designing optimal gRNAs for three case studies: tiling CRISPRbe library for BRCA1 using the base editor BE4max, tiling RNA-targeting libraries for CD46 and CD55 using CasRx, and activation of MMP7 using CRISPRa. The crisprVerse ecosystem is open-source and deployed through the Bioconductor project to facilitate its use by the CRISPR community (https://github.com/crisprVerse).', 'published': '10.1038/s41467-022-34320-7', 'node': 20809, 'link_page': 'https://www.biorxiv.org/content/10.1101/2022.04.21.488824v2?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2022.04.21.488824v2.full.pdf'}\n",
      "{'doi': '10.1101/2022.04.21.488824', 'title': 'The crisprVerse: a comprehensive Bioconductor ecosystem for the design of CRISPR guide RNAs across nucleases and technologies', 'authors': 'Hoberecht, L.; Perampalam, P.; Lun, A.; Fortin, J.-P.', 'author_corresponding': 'Jean-Philippe  Fortin', 'author_corresponding_institution': 'Genentech', 'date': '2022-09-10', 'version': '3', 'license': 'cc_by_nc_nd', 'category': 'Bioinformatics', 'jatsxml': 'https://www.biorxiv.org/content/early/2022/09/10/2022.04.21.488824.source.xml', 'abstract': 'The success of CRISPR-mediated gene perturbation studies is highly dependent on the quality of gRNAs, and several tools have been developed to enable optimal gRNA design. However, these tools are not all adaptable to the latest CRISPR modalities or nucleases, nor do they offer comprehensive annotation methods for advanced CRISPR applications. Here, we present a new ecosystem of R packages, called crispr-Verse, that enables efficient gRNA design and annotation for a multitude of CRISPR technologies. This includes CRISPR knockout (CRISPRko), CRISPR activation (CRISPRa), CRISPR interference (CRISPRi), CRISPR base editing (CRISPRbe) and CRISPR knockdown (CRISPRkd). The core package, crisprDesign, offers a comprehensive, user-friendly, and unified interface to add on- and off-target annotations via several alignment methods, rich gene and SNP annotations, and a dozen on- and off-target activity scores. These functionalities are enabled for any RNA- or DNA-targeting nucleases, including Cas9, Cas12, and Cas13. We illustrate the general applicability of our tools by designing optimal gRNAs for three case studies: tiling CRISPRbe library for BRCA1 using the base editor BE4max, tiling RNA-targeting libraries for CD46 and CD55 using CasRx, and activation of MMP7 using CRISPRa. The crisprVerse ecosystem is open-source and deployed through the Bioconductor project to facilitate its use by the CRISPR community (https://github.com/crisprVerse).', 'published': '10.1038/s41467-022-34320-7', 'node': 20810, 'link_page': 'https://www.biorxiv.org/content/10.1101/2022.04.21.488824v3?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2022.04.21.488824v3.full.pdf'}\n",
      "{'doi': '10.1101/2023.03.03.530964', 'title': 'Widespread CRISPR repeat-like RNA regulatory elements in CRISPR-Cas systems', 'authors': 'Shmakov, S. A.; Barth, Z.; Makarova, K. S.; Wolf, Y. I.; Brover, V.; Peters, J. E.; Koonin, E. V.', 'author_corresponding': 'Eugene V. Koonin', 'author_corresponding_institution': 'NIH', 'date': '2023-03-03', 'version': '1', 'license': 'cc_by_nc_nd', 'category': 'Genomics', 'jatsxml': 'https://www.biorxiv.org/content/early/2023/03/03/2023.03.03.530964.source.xml', 'abstract': 'CRISPR-cas loci typically contain CRISPR arrays with unique spacers separating direct repeats. Spacers along with portions of adjacent repeats are transcribed and processed into CRISPR(cr) RNAs that target complementary sequences (protospacers) in mobile genetic elements, resulting in cleavage of the target DNA or RNA. Additional, standalone repeats in some CRISPR-cas loci produce distinct cr-like RNAs implicated in regulatory or other functions. We developed a computational pipeline to systematically predict crRNA-like elements by scanning for standalone repeat sequences that are conserved in closely related CRISPR-cas loci. Numerous crRNA-like elements were detected in diverse CRISPR-Cas systems, mostly, of type I, but also subtype V-A. Standalone repeats often form mini-arrays containing two repeat-like sequence separated by a spacer that is partially complementary to promoter regions of cas genes, in particular cas8, or cargo genes located within CRISPR-Cas loci, such as toxins-antitoxins. We show experimentally that a mini-array from a type I-F1 CRISPR-Cas system functions as a regulatory guide. We also identified mini-arrays in bacteriophages that could abrogate CRISPR immunity by inhibiting effector expression. Thus, recruitment of CRISPR effectors for regulatory functions via spacers with partial complementarity to the target is a common feature of diverse CRISPR-Cas systems.', 'published': 'NA', 'node': 62965, 'link_page': 'https://www.biorxiv.org/content/10.1101/2023.03.03.530964v1?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2023.03.03.530964v1.full.pdf'}\n",
      "-------------------------\n",
      "CPU times: user 4.13 ms, sys: 9 µs, total: 4.14 ms\n",
      "Wall time: 11 ms\n"
     ]
    }
   ],
   "source": [
    "%%time \n",
    "\n",
    "query = {\n",
    "    \"match\": {\n",
    "        \"abstract\": \"CRISPR\"\n",
    "    }\n",
    "}\n",
    "\n",
    "results = search_data(es, index_name, query)\n",
    "\n",
    "print(\"Results: \",len(results), \"\\n\")\n",
    "# Display the results directly\n",
    "for result in results:\n",
    "    # Assuming you want to display each document's key-value pairs\n",
    "    print(result)\n",
    "\n",
    "print(\"-\" * 25)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23f16e76-5927-4760-b131-35e0a4ca47a7",
   "metadata": {},
   "source": [
    "### Term Query\n",
    "Retrieve documents where the `license` field exactly matches the specified value.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "92d8b26c-3c7c-49ee-984f-fa6eb4b90a18",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Results:  10 \n",
      "\n",
      "{'doi': '10.1101/2022.01.10.475621', 'title': 'Ecosystem size-induced environmental fluctuations affect the temporal dynamics of community assembly mechanisms', 'authors': 'Bier, R. L.; Vass, M.; Szekely, A. J.; Langenheder, S.', 'author_corresponding': 'Raven L Bier', 'author_corresponding_institution': 'Department of Ecology and Genetics/Limnology, Uppsala University, Sweden', 'date': '2022-01-11', 'version': '1', 'license': 'cc_by_nc_nd', 'category': 'Ecology', 'jatsxml': 'https://www.biorxiv.org/content/early/2022/01/11/2022.01.10.475621.source.xml', 'abstract': 'Understanding processes that determine community membership and abundance is important for many fields from theoretical community ecology to conservation. However, spatial community studies are often conducted only at a single timepoint despite the known influence of temporal variability on community assembly processes. Here we used a spatiotemporal study to determine how environmental fluctuation differences induced by mesocosm volumes (larger volumes were more stable) influence assembly processes of aquatic bacterial metacommunities along a press disturbance gradient. By combining path analysis and network approaches, we found mesocosm size categories had distinct relative influences of assembly process and environmental factors that determined spatiotemporal bacterial community composition, including dispersal and species sorting by conductivity. These processes depended on, but were not affected proportionately by, mesocosm size. Low fluctuation, large mesocosms primarily developed through the interplay of species sorting that became more important over time and transient priority effects as evidenced by more time-delayed associations. High fluctuation, small mesocosms had regular disruptions to species sorting and greater importance of ecological drift and dispersal limitation indicated by lower richness and higher taxa replacement. Together, these results emphasize that environmental fluctuations influence ecosystems over time and its impacts are modified by biotic properties intrinsic to ecosystem size.', 'published': '10.1038/s41396-022-01286-9', 'node': 7609, 'link_page': 'https://www.biorxiv.org/content/10.1101/2022.01.10.475621v1?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2022.01.10.475621v1.full.pdf'}\n",
      "{'doi': '10.1101/2022.01.10.475639', 'title': 'Maternal transmission as a symbiont sieve, and the absence of lactation in male mammals', 'authors': 'Constable, G. W. A.; Fagan, B.; Law, R.', 'author_corresponding': 'Richard  Law', 'author_corresponding_institution': 'University of York', 'date': '2022-01-11', 'version': '1', 'license': 'cc_by_nc_nd', 'category': 'Evolutionary Biology', 'jatsxml': 'https://www.biorxiv.org/content/early/2022/01/11/2022.01.10.475639.source.xml', 'abstract': 'Gut microbiomes of humans carry a complex symbiotic assemblage of microorganisms. As in all mammals, the special mode of feeding newborn infants through milk from the mammary gland enhances the opportunity for vertical transmission of the milk microbiome from parents to the gut microbiome of offspring. This has potential benefits, but it also brings with it some hazards for the host. Here we use mathematical and numerical models to demonstrate that vertical transmission from both parents would allow host populations to be invaded by microbiome elements that are deleterious. In contrast, vertical transmission, when restricted to one parent, acts as a sieve preventing the spread of such elements. We show that deleterious symbionts generate selection for uniparental transmission in host populations, and that this selective advantage is maintained in the presence of moderate horizontal transmission. Some vertical transmission from mother to infant is bound to happen in placental mammals. This paper therefore puts forward the hypothesis that the asymmetry between females and males, together with the hazards that come with biparental transmission of the milk microbiome, generate selection against male lactation in humans, and in mammals in general.', 'published': 'NA', 'node': 7610, 'link_page': 'https://www.biorxiv.org/content/10.1101/2022.01.10.475639v1?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2022.01.10.475639v1.full.pdf'}\n",
      "{'doi': '10.1101/2022.01.10.475639', 'title': 'Maternal transmission as a microbial symbiont sieve, and the absence of lactation in male mammals', 'authors': 'Fagan, B.; Constable, G. W. A.; Law, R.', 'author_corresponding': 'Richard  Law', 'author_corresponding_institution': 'University of York', 'date': '2022-08-11', 'version': '2', 'license': 'cc_by_nc_nd', 'category': 'Evolutionary Biology', 'jatsxml': 'https://www.biorxiv.org/content/early/2022/08/11/2022.01.10.475639.source.xml', 'abstract': 'Gut microbiomes of humans carry a complex symbiotic assemblage of microorganisms. As in all mammals, the special mode of feeding newborn infants through milk from the mammary gland enhances the opportunity for vertical transmission of the milk microbiome from parents to the gut microbiome of offspring. This has potential benefits, but it also brings with it some hazards for the host. Here we use mathematical and numerical models to demonstrate that vertical transmission from both parents would allow host populations to be invaded by microbiome elements that are deleterious. In contrast, vertical transmission, when restricted to one parent, acts as a sieve preventing the spread of such elements. We show that deleterious symbionts generate selection for uniparental transmission in host populations, and that this selective advantage is maintained in the presence of moderate horizontal transmission. Some vertical transmission from mother to infant is bound to happen in placental mammals. This paper therefore puts forward the hypothesis that the asymmetry between females and males, together with the hazards that come with biparental transmission of the milk microbiome, generate selection against male lactation in humans, and in mammals in general.', 'published': 'NA', 'node': 7611, 'link_page': 'https://www.biorxiv.org/content/10.1101/2022.01.10.475639v2?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2022.01.10.475639v2.full.pdf'}\n",
      "{'doi': '10.1101/2022.01.10.475641', 'title': 'Stability of the genetic structure of a wild wheat population across microhabitats over 36 years', 'authors': 'Dahan-Meir, T.; Ellis, T. J.; Mafessoni, F.; Sela, H.; Manisterski, J.; Avivi-Ragolsky, N.; Raz, A.; Feldman, M.; Anikster, Y.; Nordborg, M.; Levy, A. A.', 'author_corresponding': 'Avraham A. Levy', 'author_corresponding_institution': 'Department of Plant and Environmental Sciences, Weizmann Institute of Science; Rehovot, Israel.', 'date': '2022-05-01', 'version': '2', 'license': 'cc_by_nc_nd', 'category': 'Evolutionary Biology', 'jatsxml': 'https://www.biorxiv.org/content/early/2022/05/01/2022.01.10.475641.source.xml', 'abstract': 'Long-term genetic studies of wild populations are very scarce, but are essential for connecting ecological and population genetics models, and for understanding the dynamics of biodiversity. We present a study of a wild wheat population sampled over a 36-year period at high spatial resolution. We genotyped 832 individuals from regular sampling along transects during the course of the experiment. Genotypes were clustered into ecological microhabitats over scales of tens of metres, and this clustering was remarkably stable over the 36 generations of the study. Simulations show that it is difficult to explain this spatial and temporal stability using only limited dispersal, suggesting a role for fine-scale local adaptation to ecological parameters. Using a common-garden experiment, we showed that the genotypes found in distinct microhabitats differ phenotypically, further supporting the hypothesis of local adaptation. Our results provide a rare insight into the population genetics of a natural population over a long monitoring period.', 'published': 'NA', 'node': 7613, 'link_page': 'https://www.biorxiv.org/content/10.1101/2022.01.10.475641v2?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2022.01.10.475641v2.full.pdf'}\n",
      "{'doi': '10.1101/2022.01.10.475641', 'title': '36-year study reveals the effects of microhabitats on a wild wheat population', 'authors': 'Dahan-Meir, T.; Ellis, T. J.; Mafessoni, F.; Sela, H.; Rudich, O.; Manisterski, J.; Avivi-Ragolsky, N.; Raz, A.; Feldman, M.; Anikster, Y.; Nordborg, M.; Levy, A. A.', 'author_corresponding': 'Avraham A. Levy', 'author_corresponding_institution': 'Department of Plant and Environmental Sciences, Weizmann Institute of Science; Rehovot, Israel.', 'date': '2022-09-29', 'version': '3', 'license': 'cc_by_nc_nd', 'category': 'Evolutionary Biology', 'jatsxml': 'https://www.biorxiv.org/content/early/2022/09/29/2022.01.10.475641.source.xml', 'abstract': 'Long-term genetic studies of wild populations are very scarce, but are essential for connecting ecological and population genetics models, and for understanding the dynamics of biodiversity. We present a study of a wild wheat population sampled over a 36-year period at high spatial resolution. We genotyped 832 individuals from regular sampling along transects during the course of the experiment. Genotypes were clustered into ecological microhabitats over scales of tens of metres, and this clustering was remarkably stable over the 36 generations of the study. Simulations show that it is difficult to explain this spatial and temporal stability using only limited dispersal, suggesting a role for fine-scale local adaptation to ecological parameters. Using a common-garden experiment, we showed that the genotypes found in distinct microhabitats differ phenotypically, further supporting the hypothesis of local adaptation. Our results provide a rare insight into the population genetics of a natural population over a long monitoring period.', 'published': 'NA', 'node': 7614, 'link_page': 'https://www.biorxiv.org/content/10.1101/2022.01.10.475641v3?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2022.01.10.475641v3.full.pdf'}\n",
      "{'doi': '10.1101/2022.01.10.475641', 'title': '36-year study reveals stability of a wild wheat population across microhabitats', 'authors': 'Dahan-Meir, T.; Ellis, T. J.; Mafessoni, F.; Sela, H.; Rudich, O.; Manisterski, J.; Avivi-Ragolsky, N.; Raz, A.; Feldman, M.; Anikster, Y.; Nordborg, M.; Levy, A. A.', 'author_corresponding': 'Avraham A. Levy', 'author_corresponding_institution': 'Department of Plant and Environmental Sciences, Weizmann Institute of Science; Rehovot, Israel.', 'date': '2023-08-08', 'version': '4', 'license': 'cc_by_nc_nd', 'category': 'Evolutionary Biology', 'jatsxml': 'https://www.biorxiv.org/content/early/2023/08/08/2022.01.10.475641.source.xml', 'abstract': 'Long-term genetic studies of wild populations are very scarce, but are essential for connecting ecological and population genetics models, and for understanding the dynamics of biodiversity. We present a study of a wild wheat population sampled over a 36-year period at high spatial resolution. We genotyped 832 individuals from regular sampling along transects during the course of the experiment. Genotypes were clustered into ecological microhabitats over scales of tens of metres, and this clustering was remarkably stable over the 36 generations of the study. Simulations show that it is difficult to explain this spatial and temporal stability using only limited dispersal, suggesting a role for fine-scale local adaptation to ecological parameters. Using a common-garden experiment, we showed that the genotypes found in distinct microhabitats differ phenotypically, further supporting the hypothesis of local adaptation. Our results provide a rare insight into the population genetics of a natural population over a long monitoring period.', 'published': 'NA', 'node': 7615, 'link_page': 'https://www.biorxiv.org/content/10.1101/2022.01.10.475641v4?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2022.01.10.475641v4.full.pdf'}\n",
      "{'doi': '10.1101/2022.01.10.475153', 'title': 'Evaluation of methods incorporating biological function and GWAS summary statistics to accelerate discovery', 'authors': 'Moore, A.; Marks, J.; Quach, B. C.; Guo, Y.; Bierut, L. J.; Gaddis, N. C.; Hancock, D. B.; Page, G. P.; Johnson, E. O.', 'author_corresponding': 'Eric Otto Johnson', 'author_corresponding_institution': 'GenOmics, Bioinformatics, and Translational Research Center, Biostatistics and Epidemiology Division, RTI International, Research Triangle Park, NC 27709', 'date': '2022-01-11', 'version': '1', 'license': 'cc_by_nc_nd', 'category': 'Genetics', 'jatsxml': 'https://www.biorxiv.org/content/early/2022/01/11/2022.01.10.475153.source.xml', 'abstract': 'Where sufficiently large genome-wide association study (GWAS) samples are not currently available or feasible, methods that leverage increasing knowledge of the biological function of variants may illuminate discoveries without increasing sample size. We comprehensively evaluated 18 functional weighting methods for identifying novel associations. We assessed the performance of these methods using published results from multiple GWAS waves across each of five complex traits. Although no method achieved both high sensitivity and positive predictive value (PPV) for any trait, a subset of methods utilizing pleiotropy and expression quantitative trait loci nominated variants with high PPV (>75%) for multiple traits. Application of functionally weighting methods to enhance GWAS power for locus discovery is unlikely to circumvent the need for larger sample sizes in truly underpowered GWAS, but these results suggest that applying functional weighting to GWAS can accurately nominate additional novel loci from available samples for follow-up studies.', 'published': 'NA', 'node': 7618, 'link_page': 'https://www.biorxiv.org/content/10.1101/2022.01.10.475153v1?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2022.01.10.475153v1.full.pdf'}\n",
      "{'doi': '10.1101/2022.01.10.475652', 'title': 'Model-based hearing-restoration strategies for cochlear synaptopathy pathologies', 'authors': 'Drakopoulos, F.; Vasilkov, V.; Osses Vecchi, A.; Wartenberg, T.; Verhulst, S.', 'author_corresponding': 'Fotios  Drakopoulos', 'author_corresponding_institution': 'Ghent University, Ghent, Belgium', 'date': '2022-01-11', 'version': '1', 'license': 'cc_by_nc_nd', 'category': 'Bioengineering', 'jatsxml': 'https://www.biorxiv.org/content/early/2022/01/11/2022.01.10.475652.source.xml', 'abstract': 'It is well known that ageing and noise exposure are important causes of sensorineural hearing loss, and can result in damage of the outer hair cells or other structures of the inner ear, including synaptic damage to the auditory nerve (AN), i.e., cochlear synaptopathy (CS). Despite the suspected high prevalence of CS among people with self-reported hearing difficulties but seemingly normal hearing, conventional hearing-aid algorithms do not compensate for the functional deficits associated with CS. Here, we present and evaluate a number of auditory signal-processing strategies designed to maximally restore AN coding for listeners with CS pathologies. We evaluated our algorithms in subjects with and without suspected age-related CS to assess whether physiological and behavioural markers associated with CS can be improved. Our data show that after applying our algorithms, envelope-following responses and perceptual amplitude-modulation sensitivity were consistently enhanced in both young and older listeners. Speech-in-noise intelligibility showed small improvements after processing but mostly for young normal-hearing participants, with median improvements of up to 8.3%. Since our hearing-enhancement strategies were designed to optimally drive the AN fibres, they were able to improve temporal-envelope processing for listeners both with and without suspected CS. Our proposed algorithms can be rapidly executed and can thus extend the application range of current hearing aids and hearables, while leaving sound amplification unaffected.', 'published': '10.1016/j.heares.2022.108569', 'node': 7622, 'link_page': 'https://www.biorxiv.org/content/10.1101/2022.01.10.475652v1?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2022.01.10.475652v1.full.pdf'}\n",
      "{'doi': '10.1101/2022.01.10.475652', 'title': 'Model-based hearing-enhancement strategies for cochlear synaptopathy pathologies', 'authors': 'Drakopoulos, F.; Vasilkov, V.; Osses Vecchi, A.; Wartenberg, T.; Verhulst, S.', 'author_corresponding': 'Fotios  Drakopoulos', 'author_corresponding_institution': 'Ghent University, Ghent, Belgium', 'date': '2022-07-07', 'version': '2', 'license': 'cc_by_nc_nd', 'category': 'Bioengineering', 'jatsxml': 'https://www.biorxiv.org/content/early/2022/07/07/2022.01.10.475652.source.xml', 'abstract': 'It is well known that ageing and noise exposure are important causes of sensorineural hearing loss, and can result in damage of the outer hair cells or other structures of the inner ear, including synaptic damage to the auditory nerve (AN), i.e., cochlear synaptopathy (CS). Despite the suspected high prevalence of CS among people with self-reported hearing difficulties but seemingly normal hearing, conventional hearing-aid algorithms do not compensate for the functional deficits associated with CS. Here, we present and evaluate a number of auditory signal-processing strategies designed to maximally restore AN coding for listeners with CS pathologies. We evaluated our algorithms in subjects with and without suspected age-related CS to assess whether physiological and behavioural markers associated with CS can be improved. Our data show that after applying our algorithms, envelope-following responses and perceptual amplitude-modulation sensitivity were consistently enhanced in both young and older listeners. Speech-in-noise intelligibility showed small improvements after processing but mostly for young normal-hearing participants, with median improvements of up to 8.3%. Since our hearing-enhancement strategies were designed to optimally drive the AN fibres, they were able to improve temporal-envelope processing for listeners both with and without suspected CS. Our proposed algorithms can be rapidly executed and can thus extend the application range of current hearing aids and hearables, while leaving sound amplification unaffected.', 'published': '10.1016/j.heares.2022.108569', 'node': 7623, 'link_page': 'https://www.biorxiv.org/content/10.1101/2022.01.10.475652v2?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2022.01.10.475652v2.full.pdf'}\n",
      "{'doi': '10.1101/2022.01.10.475606', 'title': 'Establishment of a hypoxia ischemia reperfusion brain damagemodel in neonatal rats', 'authors': 'Fu, X.; Zhang, T.; Lin, W.; Jiao, M.; Zhang, Z.; Geng, J.; Wei, Q.; Qi, A.; Lin, K.; Zheng, Y.; Fang, M.; Li, P.; Lin, Z.', 'author_corresponding': 'Zhenlang  Lin', 'author_corresponding_institution': \"The Second Affiliated Hospital and Yuying Children's Hospital of Wenzhou Medical University\", 'date': '2022-01-11', 'version': '1', 'license': 'cc_by_nc_nd', 'category': 'Neuroscience', 'jatsxml': 'https://www.biorxiv.org/content/early/2022/01/11/2022.01.10.475606.source.xml', 'abstract': 'ObjectiveRice-Vannucci model has been widely used as HIE (Hypoxic ischemic encephalopathy) animal model in the past forty years, but it does not mimic reperfusion injury that occurs during HIE. The aim of the present study was to establish a new neonatal rat model by simulating hypoxia ischemia reperfusion brain damage (HIRBD) through \"common carotid artery (CCA) muscle bridge\".\\n\\nMethodsSixty 7-day-old male Sprague-Dawley rats were randomly assigned to group A (HIRBD groups, n=36), group B (Rice-Vannucci group, n=12), and group C (sham-operated group, n=12). Rats in group A were assigned to 3 subgroups (A1-A3, 12 animals/subgroup). Dynamic changes in cerebral blood flow (CBF) were evaluated by the laser speckle imaging system. The status of the CCA was observed under a stereomicroscope. Changes in body weight, gross morphology as well as pathological sections of brain tissue were examined to evaluate the feasibility of the model.\\n\\nResultsThe results indicated that CCA muscle bridge successfully blocked the CBF. CBF was restored after removal of the CCA muscle bridge in HIRBD groups. The CCA was in good condition after removing the muscle bridge, and blood supply was not affected. Changes in body weight, gross morphology and pathological sections of brain tissue indicated that ischemia reperfusion induced by the CCA muscle bridge method caused varying degrees of brain damage.\\n\\nConclusionCCA muscle bridge method is effective for establishing a reliable, stable, and reproducible neonatal rat model for study of HIRBD.', 'published': 'NA', 'node': 7626, 'link_page': 'https://www.biorxiv.org/content/10.1101/2022.01.10.475606v1?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2022.01.10.475606v1.full.pdf'}\n",
      "-------------------------\n",
      "CPU times: user 7.97 ms, sys: 0 ns, total: 7.97 ms\n",
      "Wall time: 16.7 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "query = {\n",
    "    \"term\": {\n",
    "        \"license\": {\n",
    "            \"value\": \"cc_by_nc_nd\"\n",
    "        }\n",
    "    }\n",
    "}\n",
    "\n",
    "# Execute the search\n",
    "results = search_data(es, index_name, query)\n",
    "\n",
    "print(\"Results: \",len(results), \"\\n\")\n",
    "# Display the results directly\n",
    "for result in results:\n",
    "    # Assuming you want to display each document's key-value pairs\n",
    "    print(result)\n",
    "\n",
    "print(\"-\" * 25)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46450067-4af5-443b-947c-4ab35a1dccce",
   "metadata": {},
   "source": [
    "### Range Query\n",
    "Find documents published within a specific date range.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "3f09bdab-55c6-4e56-bfab-e63fa89e3724",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Results:  10 \n",
      "\n",
      "{'doi': '10.1101/2022.01.12.475145', 'title': 'Exercise reprograms the inflammatory landscape of multiple stem cell compartments during mammalian aging', 'authors': 'Liu, L.; Kim, S.; Buckley, M. T.; Reyes, J. M.; Kang, J.; Tian, L.; Wang, T.; Lieu, A.; Mao, M.; Mateo-Rodriguez, C.; Ishak, H.; Jeong, M.; Wu, J. C.; Goodell, M. A.; Brunet, A.; Rando, T. A.', 'author_corresponding': 'Thomas A Rando', 'author_corresponding_institution': 'Stanford University', 'date': '2022-12-25', 'version': '2', 'license': 'cc_no', 'category': 'Cell Biology', 'jatsxml': 'https://www.biorxiv.org/content/early/2022/12/25/2022.01.12.475145.source.xml', 'abstract': 'Exercise has the ability to rejuvenate stem cells and improve tissue homeostasis and regeneration in aging animals. However, the cellular and molecular changes elicited by exercise have not been systematically studied across a broad range of cell types in stem cell compartments. To gain better insight into the mechanisms by which exercise affects niche and stem cell function, we subjected young and old mice to aerobic exercise and generated a single cell transcriptomic atlas of muscle, neural and hematopoietic stem cells with their niche cells and progeny.\\n\\nComplementarily, we also performed whole transcriptome analysis of single myofibers from these animals. We identified common and unique pathways that are compromised across these tissues and cell types in aged animals. We found that exercise has a rejuvenating effect on subsets of stem cells, and a profound impact in the composition and transcriptomic landscape of both circulating and tissue resident immune cells. Exercise ameliorated the upregulation of a number of inflammatory pathways as well as restored aspects of cell-cell communication within these stem cell compartments. Our study provides a comprehensive view of the coordinated responses of multiple aged stem cells and niche cells to exercise at the transcriptomic level.', 'published': '10.1016/j.stem.2023.03.016', 'node': 7735, 'link_page': 'https://www.biorxiv.org/content/10.1101/2022.01.12.475145v2?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2022.01.12.475145v2.full.pdf'}\n",
      "{'doi': '10.1101/2022.01.14.476310', 'title': 'Non-target effects of ten essential oils on the egg parasitoid Trichogramma evanescens', 'authors': 'Van Oudenhove, L.; Cazier, A.; Fillaud, M.; Lavoir, A.-V.; Fatnassi, H.; Perez, G.; Calcagno, V.', 'author_corresponding': 'Louise  Van Oudenhove', 'author_corresponding_institution': \"Institut Sophia Agrobiotech, INRAE, Universite Cote d'Azur, CNRS, France.\", 'date': '2022-12-15', 'version': '4', 'license': 'cc_by_nc_nd', 'category': 'Ecology', 'jatsxml': 'https://www.biorxiv.org/content/early/2022/12/15/2022.01.14.476310.source.xml', 'abstract': 'Essential oils (EOs) are increasingly used as biopesticides due to their insecticidal potential. This study addresses their non-target effects on a biological control agent: the egg parasitoid Trichogramma evanescens. In particular, we tested whether EOs affected parasitoid fitness either directly, by decreasing pre-imaginal survival, or indirectly, by disrupting parasitoids orientation abilities. The effect of Anise, Fennel, Sweet orange, Basil, Coriander, Oregano, Peppermint, Mugwort, Rosemary and Thyme EOs were studied on five strains of T. evanescens. Specific experimental setups were developed, and data obtained from image analysis were interpreted with phenomenological models fitted with Bayesian inference. Results highlight the fumigant toxicity of EOs on parasitoid development. Anise, Fennel, Basil, Coriander, Oregano, Peppermint and Thyme EOs are particularly toxic and drastically reduce the emergence rate of T. evanescens. Most EOs also affect parasitoid behavior: (i) Basil, Coriander, Oregano, Peppermint, Mugwort and Thyme EOs are highly repellent for naive female parasitoids; (ii) Anise and Fennel EOs can have repellent or attractive effects depending on strains; and (iii) Sweet orange, Oregano and Rosemary EOs have no detectable impact on orientation behavior. This study shows that EOs fumigation have non-target effects on egg parasitoids. This highlights the need to cautiously precise the deployment framework of biopesticides in an agroecological perspective.', 'published': '10.24072/pcjournal.212', 'node': 8123, 'link_page': 'https://www.biorxiv.org/content/10.1101/2022.01.14.476310v4?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2022.01.14.476310v4.full.pdf'}\n",
      "{'doi': '10.1101/2022.01.14.476310', 'title': 'Non-target effects of ten essential oils on the egg parasitoid Trichogramma evanescens', 'authors': 'Van Oudenhove, L.; Cazier, A.; Fillaud, M.; Lavoir, A.-V.; Fatnassi, H.; Perez, G.; Calcagno, V.', 'author_corresponding': 'Louise  Van Oudenhove', 'author_corresponding_institution': \"Institut Sophia Agrobiotech, INRAE, Universite Cote d'Azur, CNRS, France.\", 'date': '2022-12-26', 'version': '5', 'license': 'cc_by_nc_nd', 'category': 'Ecology', 'jatsxml': 'https://www.biorxiv.org/content/early/2022/12/26/2022.01.14.476310.source.xml', 'abstract': 'Essential oils (EOs) are increasingly used as biopesticides due to their insecticidal potential. This study addresses their non-target effects on a biological control agent: the egg parasitoid Trichogramma evanescens. In particular, we tested whether EOs affected parasitoid fitness either directly, by decreasing pre-imaginal survival, or indirectly, by disrupting parasitoids orientation abilities. The effect of Anise, Fennel, Sweet orange, Basil, Coriander, Oregano, Peppermint, Mugwort, Rosemary and Thyme EOs were studied on five strains of T. evanescens. Specific experimental setups were developed, and data obtained from image analysis were interpreted with phenomenological models fitted with Bayesian inference. Results highlight the fumigant toxicity of EOs on parasitoid development. Anise, Fennel, Basil, Coriander, Oregano, Peppermint and Thyme EOs are particularly toxic and drastically reduce the emergence rate of T. evanescens. Most EOs also affect parasitoid behavior: (i) Basil, Coriander, Oregano, Peppermint, Mugwort and Thyme EOs are highly repellent for naive female parasitoids; (ii) Anise and Fennel EOs can have repellent or attractive effects depending on strains; and (iii) Sweet orange, Oregano and Rosemary EOs have no detectable impact on orientation behavior. This study shows that EOs fumigation have non-target effects on egg parasitoids. This highlights the need to cautiously precise the deployment framework of biopesticides in an agroecological perspective.', 'published': '10.24072/pcjournal.212', 'node': 8124, 'link_page': 'https://www.biorxiv.org/content/10.1101/2022.01.14.476310v5?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2022.01.14.476310v5.full.pdf'}\n",
      "{'doi': '10.1101/2022.01.16.476383', 'title': 'A novel mechanism for left-right asymmetry establishment involving tissue remodeling and MyoID', 'authors': 'Lefevre, B. M.; Delvigne, M.; Vidal, J.; Courtier-Orgogozo, V.; Lang, M.', 'author_corresponding': 'Michael  Lang', 'author_corresponding_institution': 'Université de Paris, CNRS, Institut Jacques Monod, F-75013 Paris, France.', 'date': '2022-12-19', 'version': '2', 'license': 'cc_by_nc', 'category': 'Evolutionary Biology', 'jatsxml': 'https://www.biorxiv.org/content/early/2022/12/19/2022.01.16.476383.source.xml', 'abstract': 'Left-right asymmetries recurrently evolve in animals but the underlying developmental mechanisms are unknown. In most Drosophila species, the male genitalia is symmetric and undergoes a clockwise 360{degrees} rotation during development. In D. melanogaster, this tissue remodeling process is directed by class I myosins: in MyoID mutants the rotation is reversed and the genitalia remains symmetric. Males of Drosophila pachea have evolved unique left-right asymmetric genital organs and a characteristic right-sided copulation posture in the past 3-6 million years. To test if these asymmetries in morphology and behavior evolved via the recruitment of pre-existing directional cues controlled by MyoID, we used CRISPR to knockout MyoID in D. pachea. Strikingly, mutant males undergo a reverse genitalia rotation and develop mirror-image asymmetric genitalia, indicating that MyoID controls both the direction of genitalia rotation direction and morphological asymmetry. Although genital asymmetry is reversed, MyoID mutants still adopt a wild-type right-sided copulation posture. Our results show that MyoID and its asymmetry guidance role were recruited for the evolution of a novel left-right asymmetry in D. pachea and that lateralized copulation behavior is determined by other mechanisms. The evolution of a new left-right organ size asymmetry in D. pachea involved recruitment of existing directional cues.', 'published': 'NA', 'node': 8386, 'link_page': 'https://www.biorxiv.org/content/10.1101/2022.01.16.476383v2?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2022.01.16.476383v2.full.pdf'}\n",
      "{'doi': '10.1101/2022.01.17.476694', 'title': 'Catch bond models may explain how force amplifies TCR signaling and antigen discrimination', 'authors': 'Choi, H.-K.; Cong, P.; Ge, C.; Natarajan, A.; Liu, B.; Zhang, Y.; Li, K.; Rushdi, M. N.; Chen, W.; Lou, J.; Krogsgaard, M.; Zhu, C.', 'author_corresponding': 'Cheng  Zhu', 'author_corresponding_institution': 'Georgia Institute of Technology', 'date': '2022-12-24', 'version': '4', 'license': 'cc_no', 'category': 'Biophysics', 'jatsxml': 'https://www.biorxiv.org/content/early/2022/12/24/2022.01.17.476694.source.xml', 'abstract': 'Central to T cell biology, the T cell receptor (TCR) integrates forces in its triggering process upon interaction with peptide-major histocompatibility complex (pMHC)1-3. Phenotypically, forces elicit TCR catch-slip bonds with strong pMHCs but slip-only bonds with weak pMHCs4-10. While such correlation is commonly observed, the quantitative bond pattern and degree of \"catchiness\" vary. We developed two models based on the structure, elastic properties, and force-induced conformational changes of the TCR-pMHC-I/II complexes to derive from their bond characteristics more intrinsic parameters that underlie structural mechanisms, predict T cell signaling, and discriminate antigens. Applying the models to 55 datasets of 12 {beta}TCRs and their mutants interacting with corresponding pMHCs without coreceptor engagement demonstrated the ability for structural and physical parameters to quantitatively integrate and classify a broad range of bond behaviors and biological activities. Comparing to the generic two-state model for catch-slip bond that also fits the data, our models can distinguish class I from class II MHC systems and their best-fit parameters correlate with the TCR/pMHC potency to trigger T cell activation, which the generic model cannot. The models were tested by mutagenesis using structural analysis, bond profile measurement, and functional assay of a MHC and a TCR mutated to alter conformation changes. The extensive comparisons between theory and experiment provided strong validation of the models and testable hypothesis regarding specific conformational changes that control bond profiles, thereby suggesting structural mechanisms for the inner workings of the TCR mechanosensing machinery and plausible explanation of why and how force may amplify TCR signaling and antigen discrimination.', 'published': '10.1038/s41467-023-38267-1', 'node': 8423, 'link_page': 'https://www.biorxiv.org/content/10.1101/2022.01.17.476694v4?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2022.01.17.476694v4.full.pdf'}\n",
      "{'doi': '10.1101/2022.01.18.476783', 'title': 'Phenotypic Pliancy and the Breakdown of Epigenetic Polycomb Mechanisms', 'authors': 'Lambros, L.; Sella, Y.; Bergman, A.', 'author_corresponding': 'Aviv  Bergman', 'author_corresponding_institution': 'Albert Einstein College of Medicine', 'date': '2022-12-17', 'version': '3', 'license': 'cc_no', 'category': 'Systems Biology', 'jatsxml': 'https://www.biorxiv.org/content/early/2022/12/17/2022.01.18.476783.source.xml', 'abstract': 'Epigenetic regulatory mechanisms allow multicellular organisms to develop distinct specialized cell identities despite having the same total genome. Cell-fate choices are based on gene expression programs and environmental cues that cells experience during embryonic development, and are usually maintained throughout the life of the organism despite new environmental cues. The evolutionarily conserved Polycomb group (PcG) proteins form Polycomb Repressive Complexes that help orchestrate these developmental choices. Post-development, these complexes actively maintain the resulting cell fate, even in the face of environmental perturbations. Given the crucial role of these polycomb mechanisms in providing phenotypic fidelity (i.e. maintenance of cell fate), we hypothesize that their dysregulation after development will lead to decreased phenotypic fidelity allowing dysregulated cells to sustainably switch their phenotype in response to environmental changes. We call this abnormal phenotypic switching phenotypic pliancy. We introduce a general computational evolutionary model that allows us to test our systems-level phenotypic pliancy hypothesis in-silico and in a context-independent manner. We find that 1) phenotypic fidelity is an emergent systems-level property of PcG-like mechanism evolution, and 2) phenotypic pliancy is an emergent systems-level property resulting from this mechanisms dysregulation. Since there is evidence that metastatic cells behave in a phenotypically pliant manner, we hypothesize that progression to metastasis is driven by the emergence of phenotypic pliancy in cancer cells as a result of PcG mechanism dysregulation. We corroborate our hypothesis using single-cell RNA-sequencing data from metastatic cancers. We find that metastatic cancer cells are phenotypically pliant in the same manner as predicted by our model.\\n\\nSignificance StatementWe introduce the concept of cellular phenotypic pliancy- sustained abnormal phenotypic switching in response to environmental changes- and demonstrate that such behavior can be caused by dysregulation of Polycomb mechanisms. To overcome the incomplete knowledge about these mechanisms in higher organisms, we develop an abstract computational model to study the emergence of phenotypic pliancy from a general systems-level view by confirming our hypothesis over a wide range of simulated gene-regulatory networks and Polycomb mechanism patterns. We corroborate our hypothesis and model predictions using single-cell RNA-seq metastatic cancer datasets. Our hypothesis has the potential to shed light on a general phenomenon for complex diseases where abnormal phenotypic switching is relevant.', 'published': '10.1371/journal.pcbi.1010889', 'node': 8751, 'link_page': 'https://www.biorxiv.org/content/10.1101/2022.01.18.476783v3?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2022.01.18.476783v3.full.pdf'}\n",
      "{'doi': '10.1101/2022.01.20.477133', 'title': 'Outbreak.info Research Library: A standardized, searchable platform to discover and explore COVID-19 resources', 'authors': 'Tsueng, G.; Mullen, J.; Alkuzweny, M.; Cano, M. A.; Rush, B.; Haag, E.; Outbreak Curators,  ; Lin, J.; Welzel, D. J.; Zhou, X.; Qian, Z.; Latif, A. A.; Hufbauer, E.; Zeller, M.; Andersen, K. G.; Wu, C.; Su, A. I.; Gangavarapu, K.; Hughes, L. D.', 'author_corresponding': 'Ginger  Tsueng', 'author_corresponding_institution': 'The Scripps Research institute', 'date': '2022-12-07', 'version': '5', 'license': 'cc_by', 'category': 'Bioinformatics', 'jatsxml': 'https://www.biorxiv.org/content/early/2022/12/07/2022.01.20.477133.source.xml', 'abstract': 'To combat the ongoing COVID-19 pandemic, scientists have been conducting research at breakneck speeds, producing over 52,000 peer-reviewed articles within the first year. To address the challenge in tracking the vast amount of new research located in separate repositories, we developed outbreak.info Research Library, a standardized, searchable interface of COVID-19 and SARS-CoV-2 resources. Unifying metadata from sixteen repositories, we assembled a collection of over 350,000 publications, clinical trials, datasets, protocols, and other resources as of October 2022. We used a rigorous schema to enforce consistency across different sources and resource types and linked related resources. Researchers can quickly search the latest research across data repositories, regardless of resource type or repository location, via a search interface, public API, and R package. Finally, we discuss the challenges inherent in combining metadata from scattered and heterogeneous resources and provide recommendations to streamline this process to aid scientific research.', 'published': '10.1038/s41592-023-01770-w', 'node': 8868, 'link_page': 'https://www.biorxiv.org/content/10.1101/2022.01.20.477133v5?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2022.01.20.477133v5.full.pdf'}\n",
      "{'doi': '10.1101/2022.01.19.476871', 'title': 'A Population Response Model of Ensemble Coding', 'authors': 'Utochkin, I. S.; Choi, J.; Chong, S. C.', 'author_corresponding': 'Igor S. Utochkin', 'author_corresponding_institution': 'HSE University', 'date': '2022-12-17', 'version': '2', 'license': 'cc_by_nc', 'category': 'Neuroscience', 'jatsxml': 'https://www.biorxiv.org/content/early/2022/12/17/2022.01.19.476871.source.xml', 'abstract': 'Ensemble representations have been considered as one of the strategies that the visual system adopts to cope with its limited capacity. Thus, they include various statistical summaries such as mean, variance, and distributional properties and are formed over many stages of visual processing. The current study proposes a population coding model of ensemble perception to provide a theoretical and computational framework for these various facets of ensemble perception. The proposed model consists of a simple feature layer and a pooling layer. We assumed ensemble representations as population responses in the pooling layer and decoded various statistical properties from population responses. Our model successfully predicted averaging performance in orientation, size, color, and motion direction across different tasks. Furthermore, it predicted variance discrimination performance and the priming effects of feature distributions. Finally, it explained the well-known variance and set size effects and has a potential for explaining the adaptation and clustering effects.', 'published': '10.1037/rev0000426', 'node': 9001, 'link_page': 'https://www.biorxiv.org/content/10.1101/2022.01.19.476871v2?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2022.01.19.476871v2.full.pdf'}\n",
      "{'doi': '10.1101/2022.01.21.477203', 'title': 'A common venomous ancestor? Prevalent bee venom genes evolved before the aculeate stinger while few major toxins are bee-specific', 'authors': 'Koludarov, I.; Velasque, M.; Timm, T.; Greve, C.; Hamadou, A. B.; Gupta, D. K.; Lochnit, G.; Heinzinger, M.; Vilcinskas, A.; Gloag, R.; Harpur, B. A.; Podsiadlowski, L.; Rost, B.; Jackson, T.; Dutertre, S.; Stolle, E.; Von Reumont, B. M.', 'author_corresponding': 'Bjoern Marcus Von Reumont', 'author_corresponding_institution': 'LOEWE Centre for Biodiversity Genomics', 'date': '2022-12-13', 'version': '2', 'license': 'cc_by_nc', 'category': 'Evolutionary Biology', 'jatsxml': 'https://www.biorxiv.org/content/early/2022/12/13/2022.01.21.477203.source.xml', 'abstract': 'Venoms, which have evolved numerous times in animals, are ideal models of convergent trait evolution. However, detailed genomic studies of toxin-encoding genes exist for only a few animal groups. The hyper-diverse hymenopteran insects are the most speciose venomous clade, but investigation of the origin of their venom genes has been largely neglected. Utilising a combination of genomic and proteo-transcriptomic data, we investigated the origin of 11 toxin genes in 29 published and three new hymenopteran genomes and compiled an up-to-date list of prevalent bee venom proteins. Observed patterns indicate that bee venom genes predominantly originate through single gene co-option with gene duplication contributing to subsequent diversification. Most Hymenoptera venom genes are shared by all members of the clade and only melittin and the new venom protein family anthophilin1 appear unique to the bee lineage. Most venom proteins thus predate the mega-radiation of hymenopterans and the evolution of the aculeate stinger.', 'published': 'NA', 'node': 9126, 'link_page': 'https://www.biorxiv.org/content/10.1101/2022.01.21.477203v2?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2022.01.21.477203v2.full.pdf'}\n",
      "{'doi': '10.1101/2022.01.21.477229', 'title': 'Chronic demyelination of rabbit lesions is attributable to failed oligodendrocyte progenitor cell repopulation.', 'authors': 'Cooper, J. J. M.; Polanco, J. J.; Saraswat, D.; Peirick, J.; Seidl, A.; Li, Y.; Ma, D.; Sim, F.', 'author_corresponding': 'Fraser  Sim', 'author_corresponding_institution': 'SUNY University at Buffalo', 'date': '2022-12-19', 'version': '3', 'license': 'cc_by_nc_nd', 'category': 'Neuroscience', 'jatsxml': 'https://www.biorxiv.org/content/early/2022/12/19/2022.01.21.477229.source.xml', 'abstract': 'The failure of remyelination in the human CNS contributes to axonal injury and disease progression in multiple sclerosis (MS). In contrast to regions of chronic demyelination in the human brain, remyelination in murine models is preceded by abundant oligodendrocyte progenitor cell (OPC) repopulation, such that OPC density within regions of demyelination far exceeds that of normal white matter. As such, we hypothesized that efficient OPC repopulation was a prerequisite of successful remyelination, and that increased lesion volume may contribute to the failure of OPC repopulation in human brain. In this study, we characterized the pattern of OPC activation and proliferation following induction of lysolecithin-induced chronic demyelination in adult rabbits. The density of OPCs never exceeded that of normal white matter and oligodendrocyte density did not recover even at 6 months post-injection. Rabbit OPC recruitment in large lesions was further characterized by chronic Sox2 expression in OPCs located in the lesion core and upregulation of quiescence-associated Prrx1 mRNA at the lesion border. Surprisingly, when small rabbit lesions of equivalent size to mouse were induced, they too exhibited reduced OPC repopulation. However, small lesions were distinct from large lesions as they displayed an almost complete lack of OPC proliferation following demyelination. These differences in the response to demyelination suggest that both volume dependent and species-specific mechanisms are critical in the regulation of OPC proliferation and lesion repopulation and suggest that alternate models will be necessary to fully understand the mechanisms that contribute to failed remyelination in MS.\\n\\nMain Points- Chronic demyelination in the rabbit CNS was associated with reduced OPC repopulation.\\n- Quiescent OPCs accumulated around the edge of rabbit lesions.\\n- OPC and oligodendrocyte repopulation was reduced in rabbit regardless of lesion volume.', 'published': '10.1002/glia.24324', 'node': 9175, 'link_page': 'https://www.biorxiv.org/content/10.1101/2022.01.21.477229v3?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2022.01.21.477229v3.full.pdf'}\n",
      "-------------------------\n",
      "CPU times: user 5.69 ms, sys: 0 ns, total: 5.69 ms\n",
      "Wall time: 16.5 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "query = {\n",
    "    \"range\": {\n",
    "        \"date\": {\n",
    "            \"gte\": \"2022-12-01\",\n",
    "            \"lte\": \"2022-12-31\"\n",
    "        }\n",
    "    }\n",
    "}\n",
    "\n",
    "# Execute the search\n",
    "results = search_data(es, index_name, query)\n",
    "\n",
    "print(\"Results: \",len(results), \"\\n\")\n",
    "# Display the results directly\n",
    "for result in results:\n",
    "    # Assuming you want to display each document's key-value pairs\n",
    "    print(result)\n",
    "\n",
    "print(\"-\" * 25)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef58edd1-0433-4e3f-ad0d-904f629f79b5",
   "metadata": {},
   "source": [
    "\n",
    "### Bool Query\n",
    "Combine multiple search criteria. For example, search for documents by authors in a specific category and with a specific license.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "f5d24f63-dfd0-4491-85c5-5592c6818adc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Results:  10 \n",
      "\n",
      "{'doi': '10.1101/2022.10.06.511167', 'title': 'High selectivity of frequency induced transcriptional responses', 'authors': 'Givre, A.; Colman-Lerner, A.; Ponce-Dawson, S.', 'author_corresponding': 'Silvina Ponce-Dawson', 'author_corresponding_institution': 'School of Natural and Exact Sciences, University of Buenos Aires', 'date': '2022-10-07', 'version': '1', 'license': 'cc_by_nc_nd', 'category': 'Systems Biology', 'jatsxml': 'https://www.biorxiv.org/content/early/2022/10/07/2022.10.06.511167.source.xml', 'abstract': 'Cells continuously interact with their environment, detect its changes and generate responses accordingly. This requires interpreting the variations and, in many occasions, producing changes in gene expression. In this paper we use information theory and a simple transcription model to analyze the extent to which the resulting gene expression is able to identify and assess the intensity of extracellular stimuli when they are encoded in the amplitude, duration or frequency of a transcription factors nuclear concentration. We find that the maximal information transmission is, for the three codifications, ~ 1.5 - 1.8 bits, i.e., approximately 3 ranges of input strengths can be distinguished in all cases. The types of promoters that yield maximum transmission for the three modes are all similarly fast and have a high activation threshold. The three input modulation modes differ, however, in the sensitivity to changes in the parameters that characterize the promoters, with frequency modulation being the most sensitive and duration modulation, the least. This turns out to be key for signal identification. Namely, we show that, because of this sensitivity difference, it is possible to find promoter parameters that yield an information transmission within 90% of its maximum value for duration or amplitude modulation and less than 1 bit for frequency modulation. The reverse situation cannot be found within the framework of a single promoter transcription model. This means that pulses of transcription factors in the nucleus can selectively activate the promoter that is tuned to respond to frequency modulations while prolonged nuclear accumulation would activate several promoters at the same time. Thus, frequency modulation is better suited than the other encoding modes to allow the identification of external stimuli without requiring other mediators of the transduction.', 'published': 'NA', 'node': 43301, 'link_page': 'https://www.biorxiv.org/content/10.1101/2022.10.06.511167v1?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2022.10.06.511167v1.full.pdf'}\n",
      "{'doi': '10.1101/2022.10.06.511167', 'title': 'High selectivity of frequency induced transcriptional responses', 'authors': 'Givre, A.; Colman-Lerner, A.; Ponce-Dawson, S.', 'author_corresponding': 'Silvina Ponce-Dawson', 'author_corresponding_institution': 'School of Natural and Exact Sciences, University of Buenos Aires', 'date': '2022-10-07', 'version': '1', 'license': 'cc_by_nc_nd', 'category': 'Systems Biology', 'jatsxml': 'https://www.biorxiv.org/content/early/2022/10/07/2022.10.06.511167.source.xml', 'abstract': 'Cells continuously interact with their environment, detect its changes and generate responses accordingly. This requires interpreting the variations and, in many occasions, producing changes in gene expression. In this paper we use information theory and a simple transcription model to analyze the extent to which the resulting gene expression is able to identify and assess the intensity of extracellular stimuli when they are encoded in the amplitude, duration or frequency of a transcription factors nuclear concentration. We find that the maximal information transmission is, for the three codifications, ~ 1.5 - 1.8 bits, i.e., approximately 3 ranges of input strengths can be distinguished in all cases. The types of promoters that yield maximum transmission for the three modes are all similarly fast and have a high activation threshold. The three input modulation modes differ, however, in the sensitivity to changes in the parameters that characterize the promoters, with frequency modulation being the most sensitive and duration modulation, the least. This turns out to be key for signal identification. Namely, we show that, because of this sensitivity difference, it is possible to find promoter parameters that yield an information transmission within 90% of its maximum value for duration or amplitude modulation and less than 1 bit for frequency modulation. The reverse situation cannot be found within the framework of a single promoter transcription model. This means that pulses of transcription factors in the nucleus can selectively activate the promoter that is tuned to respond to frequency modulations while prolonged nuclear accumulation would activate several promoters at the same time. Thus, frequency modulation is better suited than the other encoding modes to allow the identification of external stimuli without requiring other mediators of the transduction.', 'published': 'NA', 'node': 43301, 'link_page': 'https://www.biorxiv.org/content/10.1101/2022.10.06.511167v1?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2022.10.06.511167v1.full.pdf'}\n",
      "{'doi': '10.1101/001768', 'title': 'Characterization of cell-to-cell variation in nuclear transport rates and identification of its sources', 'authors': 'Durrieu, L.; Bush, A.; Grande, A.; Johansson, R.; Janzen, D. L. I.; Gollvik, M.; Katz, A.; Cedersund, G.; Colman-Lerner, A.', 'author_corresponding': 'Alejandro  Colman-Lerner', 'author_corresponding_institution': 'IFIByNE, DFBMC, FCEN, UBA, Buenos Aires, Argentine', 'date': '2022-06-24', 'version': '2', 'license': 'cc_by_nc_nd', 'category': 'Systems Biology', 'jatsxml': 'https://www.biorxiv.org/content/early/2022/06/24/001768.source.xml', 'abstract': 'Nuclear transport is an essential part of eukaryotic cell function. Several assays exist to measure the rate of this process, but not at the single-cell level. Here, we developed a fluorescent recovery after photobleaching (FRAP)- based method to determine nuclear import and export rates independently in individual live cells. To overcome the inherent noise of single-cell measurements, we performed sequential FRAPs on the same cell. We found large cell-to-cell variation in transport rates within isogenic yeast populations. Our data suggest that a main determinant of this heterogeneity may be variability in the number of nuclear pore complexes (NPCs). For passive transport, this component explained most of the variability. Actively transported proteins were influenced by variability in additional components, including general factors such as the Ran-GTP gradient as well as specific regulators of the export rate. By considering mother-daughter pairs, we showed that mitotic segregation of the transport machinery is too noisy to control cellular inheritance. Finally, we studied mother-daughter cell asymmetry in the localization of the transcription factor Ace2, which is specifically concentrated in daughter cell nuclei. We found that this asymmetry is the outcome of a higher ratio of import rate to export rate in daughters. Interestingly, rather than reduced export in the daughter cell, as previously hypothesized, rates of both import and export are faster in daughter cells than in mother cells, but the magnitude of increase is greater for import. These results shed light into cell-to-cell variation in cellular dynamics and its sources.', 'published': '10.1016/j.isci.2022.105906', 'node': 1, 'link_page': 'https://www.biorxiv.org/content/10.1101/001768v2?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/001768v2.full.pdf'}\n",
      "{'doi': '10.1101/001768', 'title': 'Characterization of cell-to-cell variation in nuclear transport rates and identification of its sources', 'authors': 'Durrieu, L.; Bush, A.; Grande, A.; Johansson, R.; Janzen, D. L. I.; Gollvik, M.; Katz, A.; Cedersund, G.; Colman-Lerner, A.', 'author_corresponding': 'Alejandro  Colman-Lerner', 'author_corresponding_institution': 'IFIByNE, DFBMC, FCEN, UBA, Buenos Aires, Argentine', 'date': '2022-06-24', 'version': '2', 'license': 'cc_by_nc_nd', 'category': 'Systems Biology', 'jatsxml': 'https://www.biorxiv.org/content/early/2022/06/24/001768.source.xml', 'abstract': 'Nuclear transport is an essential part of eukaryotic cell function. Several assays exist to measure the rate of this process, but not at the single-cell level. Here, we developed a fluorescent recovery after photobleaching (FRAP)- based method to determine nuclear import and export rates independently in individual live cells. To overcome the inherent noise of single-cell measurements, we performed sequential FRAPs on the same cell. We found large cell-to-cell variation in transport rates within isogenic yeast populations. Our data suggest that a main determinant of this heterogeneity may be variability in the number of nuclear pore complexes (NPCs). For passive transport, this component explained most of the variability. Actively transported proteins were influenced by variability in additional components, including general factors such as the Ran-GTP gradient as well as specific regulators of the export rate. By considering mother-daughter pairs, we showed that mitotic segregation of the transport machinery is too noisy to control cellular inheritance. Finally, we studied mother-daughter cell asymmetry in the localization of the transcription factor Ace2, which is specifically concentrated in daughter cell nuclei. We found that this asymmetry is the outcome of a higher ratio of import rate to export rate in daughters. Interestingly, rather than reduced export in the daughter cell, as previously hypothesized, rates of both import and export are faster in daughter cells than in mother cells, but the magnitude of increase is greater for import. These results shed light into cell-to-cell variation in cellular dynamics and its sources.', 'published': '10.1016/j.isci.2022.105906', 'node': 1, 'link_page': 'https://www.biorxiv.org/content/10.1101/001768v2?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/001768v2.full.pdf'}\n",
      "{'doi': '10.1101/2023.06.22.546056', 'title': 'Boosting class II HLA epitope presentation to T cells with endoplasmic reticulum transmembrane domain fusion proteins', 'authors': 'Pinamonti, V.; Mellado Fuentes, A.; Schmid, T.; Cetin, M.; Lerner, T.; Hernandez Hernandez, M.; Lindner, J. M.', 'author_corresponding': 'John M Lindner', 'author_corresponding_institution': 'BioMed X Institute', 'date': '2023-06-22', 'version': '1', 'license': 'cc_by_nc_nd', 'category': 'Molecular Biology', 'jatsxml': 'https://www.biorxiv.org/content/early/2023/06/22/2023.06.22.546056.source.xml', 'abstract': 'Antigen processing and presentation are crucial for T-cell receptor engagement and potent adaptive immune responses. While establishing an antigen presentation system to identify TCR-epitope pairs from large libraries of genetically templated, ectopically expressed antigens, we observed a longitudinal suppression of antigen processing in B cells, which we call antigen processing attenuation (APA). Beyond its potential role in regulating B cell-enforced peripheral tolerance, APA represents a bottleneck to sustained antigen presentation in screening and vaccination approaches. To overcome APA we screened several pathways which might enhance processing. Notably, we identified several sequences derived from transmembrane domains (TM) of endoplasmic reticulum-resident proteins which promote sustained minigene-encoded antigen presentation. In particular, the first TM domain of SYT6, a protein associated with synaptic vesicles, greatly enhances processing and ameliorates APA. These findings suggest a new entry point into antigen processing pathways, enabling the discovery of novel immune epitopes and improving T-cell based immunization strategies.', 'published': 'NA', 'node': 77999, 'link_page': 'https://www.biorxiv.org/content/10.1101/2023.06.22.546056v1?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2023.06.22.546056v1.full.pdf'}\n",
      "{'doi': '10.1101/2023.06.22.546056', 'title': 'Boosting class II HLA epitope presentation to T cells with endoplasmic reticulum transmembrane domain fusion proteins', 'authors': 'Pinamonti, V.; Mellado Fuentes, A.; Schmid, T.; Cetin, M.; Lerner, T.; Hernandez Hernandez, M.; Lindner, J. M.', 'author_corresponding': 'John M Lindner', 'author_corresponding_institution': 'BioMed X Institute', 'date': '2023-06-22', 'version': '1', 'license': 'cc_by_nc_nd', 'category': 'Molecular Biology', 'jatsxml': 'https://www.biorxiv.org/content/early/2023/06/22/2023.06.22.546056.source.xml', 'abstract': 'Antigen processing and presentation are crucial for T-cell receptor engagement and potent adaptive immune responses. While establishing an antigen presentation system to identify TCR-epitope pairs from large libraries of genetically templated, ectopically expressed antigens, we observed a longitudinal suppression of antigen processing in B cells, which we call antigen processing attenuation (APA). Beyond its potential role in regulating B cell-enforced peripheral tolerance, APA represents a bottleneck to sustained antigen presentation in screening and vaccination approaches. To overcome APA we screened several pathways which might enhance processing. Notably, we identified several sequences derived from transmembrane domains (TM) of endoplasmic reticulum-resident proteins which promote sustained minigene-encoded antigen presentation. In particular, the first TM domain of SYT6, a protein associated with synaptic vesicles, greatly enhances processing and ameliorates APA. These findings suggest a new entry point into antigen processing pathways, enabling the discovery of novel immune epitopes and improving T-cell based immunization strategies.', 'published': 'NA', 'node': 77999, 'link_page': 'https://www.biorxiv.org/content/10.1101/2023.06.22.546056v1?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2023.06.22.546056v1.full.pdf'}\n",
      "{'doi': '10.1101/2023.04.28.538672', 'title': 'Neoepitope-specific vaccination of a patient with diffuse midline glioma targeting H3K27M induces polyclonal B and T cell responses across diverse HLA alleles', 'authors': 'Boschert, T.; Kromer, K.; Lerner, T.; Lindner, K.; Haltenhof, G.; Tan, C. L.; Jahne, K.; Poschke, I.; Bunse, L.; Grassl, N.; Mildenberger, I.; Sahm, K.; Platten, M.; Lindner, J. M.; Green, E. W.', 'author_corresponding': '', 'author_corresponding_institution': '', 'date': '2023-04-29', 'version': '1', 'license': 'cc_by_nc_nd', 'category': 'Cancer Biology', 'jatsxml': 'https://www.biorxiv.org/content/early/2023/04/29/2023.04.28.538672.source.xml', 'abstract': 'H3K27M, a driver mutation with T- and B-cell neoepitope characteristics, defines an aggressive subtype of diffuse glioma with poor survival. We functionally dissect the immune response of one patient who was treated with an H3K27M peptide vaccine and subsequently entered complete remission. The vaccine robustly expanded class II HLA-restricted peripheral H3K27M-specific T cells. Using functional assays, we characterized 34 clonally unique H3K27M-reactive T cell receptors and identified critical, conserved motifs in their CDR3 regions. Using detailed HLA mapping, we further demonstrate that diverse HLA-DQ, and -DR alleles present immunogenic H3K27M epitopes. Furthermore, we identified and profiled H3K27M-reactive B cell receptors from activated B cells in the cerebrospinal fluid. Our results uncover the breadth of the adaptive immune response against a shared clonal neoantigen across multiple HLA allelotypes and support the use of class II-restricted peptide vaccines to stimulate tumor-specific T and B cells harboring receptors with therapeutic potential.', 'published': 'NA', 'node': 70772, 'link_page': 'https://www.biorxiv.org/content/10.1101/2023.04.28.538672v1?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2023.04.28.538672v1.full.pdf'}\n",
      "{'doi': '10.1101/2023.04.28.538672', 'title': 'Neoepitope-specific vaccination of a patient with diffuse midline glioma targeting H3K27M induces polyclonal B and T cell responses across diverse HLA alleles', 'authors': 'Boschert, T.; Kromer, K.; Lerner, T.; Lindner, K.; Haltenhof, G.; Tan, C. L.; Jahne, K.; Poschke, I.; Bunse, L.; Grassl, N.; Mildenberger, I.; Sahm, K.; Platten, M.; Lindner, J. M.; Green, E. W.', 'author_corresponding': '', 'author_corresponding_institution': '', 'date': '2023-04-29', 'version': '1', 'license': 'cc_by_nc_nd', 'category': 'Cancer Biology', 'jatsxml': 'https://www.biorxiv.org/content/early/2023/04/29/2023.04.28.538672.source.xml', 'abstract': 'H3K27M, a driver mutation with T- and B-cell neoepitope characteristics, defines an aggressive subtype of diffuse glioma with poor survival. We functionally dissect the immune response of one patient who was treated with an H3K27M peptide vaccine and subsequently entered complete remission. The vaccine robustly expanded class II HLA-restricted peripheral H3K27M-specific T cells. Using functional assays, we characterized 34 clonally unique H3K27M-reactive T cell receptors and identified critical, conserved motifs in their CDR3 regions. Using detailed HLA mapping, we further demonstrate that diverse HLA-DQ, and -DR alleles present immunogenic H3K27M epitopes. Furthermore, we identified and profiled H3K27M-reactive B cell receptors from activated B cells in the cerebrospinal fluid. Our results uncover the breadth of the adaptive immune response against a shared clonal neoantigen across multiple HLA allelotypes and support the use of class II-restricted peptide vaccines to stimulate tumor-specific T and B cells harboring receptors with therapeutic potential.', 'published': 'NA', 'node': 70772, 'link_page': 'https://www.biorxiv.org/content/10.1101/2023.04.28.538672v1?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2023.04.28.538672v1.full.pdf'}\n",
      "{'doi': '10.1101/2022.10.03.510647', 'title': 'Carm1 regulates the speed of C/EBPa-induced transdifferentiation by a cofactor stealing mechanism', 'authors': 'Garcia, G. T.; Kowenz-Leutz, E.; Tian, T. V.; Klonizakis, A.; Lerner, J.; De Andres-Aguayo, L.; Berenguer, C.; Carmona, M. P.; Casadesus, M. V.; Bulteau, R.; Francesconi, M.; Peiro, S.; Zaret, K. S.; Leutz, A.; Graf, T.', 'author_corresponding': 'Achim Leutz', 'author_corresponding_institution': 'MDC, Berlin', 'date': '2022-10-04', 'version': '2', 'license': 'cc_by_nc_nd', 'category': 'Cell Biology', 'jatsxml': 'https://www.biorxiv.org/content/early/2022/10/04/2022.10.03.510647.source.xml', 'abstract': 'Cell fate decisions are driven by lineage-restricted transcription factors but how they are regulated is incompletely understood. The C/EBP-induced B cell to macrophage transdifferentiation (BMT) is a powerful system to address this question. Here we describe that C/EBP with a single arginine mutation (C/EBPR35A) induces a dramatically accelerated BMT in mouse and human cells. Changes in the expression of lineage-restricted genes occur as early as within 1 hour compared to 18 hours with the wild type. Mechanistically C/EBPR35A exhibits an increased affinity for PU.1, a bi-lineage transcription factor required for C/EBP-induced BMT. The complex induces more rapid chromatin accessibility changes and an enhanced relocation (stealing) of PU.1 from B cell to myeloid gene regulatory elements. Arginine 35 is methylated by Carm1 and inhibition of the enzyme accelerates BMT, similar to the mutant. Our data suggest that the relative proportions of methylated and unmethylated C/EBP in a bipotent progenitor can determine the velocity of cell fate choice and lineage directionality.', 'published': '10.7554/elife.83951', 'node': 42510, 'link_page': 'https://www.biorxiv.org/content/10.1101/2022.10.03.510647v2?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2022.10.03.510647v2.full.pdf'}\n",
      "{'doi': '10.1101/2022.10.03.510647', 'title': 'Carm1 regulates the speed of C/EBPa-induced transdifferentiation by a cofactor stealing mechanism', 'authors': 'Garcia, G. T.; Kowenz-Leutz, E.; Tian, T. V.; Klonizakis, A.; Lerner, J.; De Andres-Aguayo, L.; Berenguer, C.; Carmona, M. P.; Casadesus, M. V.; Bulteau, R.; Francesconi, M.; Peiro, S.; Zaret, K. S.; Leutz, A.; Graf, T.', 'author_corresponding': 'Achim Leutz', 'author_corresponding_institution': 'MDC, Berlin', 'date': '2022-10-04', 'version': '2', 'license': 'cc_by_nc_nd', 'category': 'Cell Biology', 'jatsxml': 'https://www.biorxiv.org/content/early/2022/10/04/2022.10.03.510647.source.xml', 'abstract': 'Cell fate decisions are driven by lineage-restricted transcription factors but how they are regulated is incompletely understood. The C/EBP-induced B cell to macrophage transdifferentiation (BMT) is a powerful system to address this question. Here we describe that C/EBP with a single arginine mutation (C/EBPR35A) induces a dramatically accelerated BMT in mouse and human cells. Changes in the expression of lineage-restricted genes occur as early as within 1 hour compared to 18 hours with the wild type. Mechanistically C/EBPR35A exhibits an increased affinity for PU.1, a bi-lineage transcription factor required for C/EBP-induced BMT. The complex induces more rapid chromatin accessibility changes and an enhanced relocation (stealing) of PU.1 from B cell to myeloid gene regulatory elements. Arginine 35 is methylated by Carm1 and inhibition of the enzyme accelerates BMT, similar to the mutant. Our data suggest that the relative proportions of methylated and unmethylated C/EBP in a bipotent progenitor can determine the velocity of cell fate choice and lineage directionality.', 'published': '10.7554/elife.83951', 'node': 42510, 'link_page': 'https://www.biorxiv.org/content/10.1101/2022.10.03.510647v2?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2022.10.03.510647v2.full.pdf'}\n",
      "-------------------------\n",
      "CPU times: user 3.25 ms, sys: 0 ns, total: 3.25 ms\n",
      "Wall time: 9.38 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "query = {\n",
    "    \"bool\": {\n",
    "        \"must\": [\n",
    "            {\"match\": {\"authors\": \"Colman-Lerner\"}},\n",
    "            {\"match\": {\"category\": \"Systems Biology\"}},\n",
    "            {\"term\": {\"license\": \"cc_by_nc_nd\"}}\n",
    "        ]\n",
    "    }\n",
    "}\n",
    "\n",
    "# Execute the search\n",
    "results = search_data(es, index_name, query)\n",
    "\n",
    "print(\"Results: \",len(results), \"\\n\")\n",
    "# Display the results directly\n",
    "for result in results:\n",
    "    # Assuming you want to display each document's key-value pairs\n",
    "    print(result)\n",
    "\n",
    "print(\"-\" * 25)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "960c8f80-af34-4ab7-8099-0d59a81af22d",
   "metadata": {},
   "source": [
    "### Match phrase Query\n",
    "Use a match_phrase to search for documents with titles that contain specific patterns.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "439127cf-21d8-49a0-a802-c950ff038d6f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Results:  6 \n",
      "\n",
      "{'doi': '10.1101/2022.10.03.510647', 'title': 'Carm1 regulates the speed of C/EBPa-induced transdifferentiation by a cofactor stealing mechanism', 'authors': 'Garcia, G. T.; Kowenz-Leutz, E.; Tian, T. V.; Klonizakis, A.; Lerner, J.; De Andres-Aguayo, L.; Berenguer, C.; Carmona, M. P.; Casadesus, M. V.; Bulteau, R.; Francesconi, M.; Leutz, A.; Zaret, K. S.; Zaret, K. S.; Peiro, S.', 'author_corresponding': 'Achim Leutz', 'author_corresponding_institution': 'MDC, Berlin', 'date': '2022-10-04', 'version': '1', 'license': 'cc_by_nc_nd', 'category': 'Cell Biology', 'jatsxml': 'https://www.biorxiv.org/content/early/2022/10/04/2022.10.03.510647.source.xml', 'abstract': 'Cell fate decisions are driven by lineage-restricted transcription factors but how they are regulated is incompletely understood. The C/EBP-induced B cell to macrophage transdifferentiation (BMT) is a powerful system to address this question. Here we describe that C/EBP with a single arginine mutation (C/EBPR35A) induces a dramatically accelerated BMT in mouse and human cells. Changes in the expression of lineage-restricted genes occur as early as within 1 hour compared to 18 hours with the wild type. Mechanistically C/EBPR35A exhibits an increased affinity for PU.1, a bi-lineage transcription factor required for C/EBP-induced BMT. The complex induces more rapid chromatin accessibility changes and an enhanced relocation (stealing) of PU.1 from B cell to myeloid gene regulatory elements. Arginine 35 is methylated by Carm1 and inhibition of the enzyme accelerates BMT, similar to the mutant. Our data suggest that the relative proportions of methylated and unmethylated C/EBP in a bipotent progenitor can determine the velocity of cell fate choice and lineage directionality.', 'published': '10.7554/elife.83951', 'node': 42509, 'link_page': 'https://www.biorxiv.org/content/10.1101/2022.10.03.510647v1?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2022.10.03.510647v1.full.pdf'}\n",
      "{'doi': '10.1101/2022.10.03.510647', 'title': 'Carm1 regulates the speed of C/EBPa-induced transdifferentiation by a cofactor stealing mechanism', 'authors': 'Garcia, G. T.; Kowenz-Leutz, E.; Tian, T. V.; Klonizakis, A.; Lerner, J.; De Andres-Aguayo, L.; Berenguer, C.; Carmona, M. P.; Casadesus, M. V.; Bulteau, R.; Francesconi, M.; Peiro, S.; Zaret, K. S.; Leutz, A.; Graf, T.', 'author_corresponding': 'Achim Leutz', 'author_corresponding_institution': 'MDC, Berlin', 'date': '2022-10-04', 'version': '2', 'license': 'cc_by_nc_nd', 'category': 'Cell Biology', 'jatsxml': 'https://www.biorxiv.org/content/early/2022/10/04/2022.10.03.510647.source.xml', 'abstract': 'Cell fate decisions are driven by lineage-restricted transcription factors but how they are regulated is incompletely understood. The C/EBP-induced B cell to macrophage transdifferentiation (BMT) is a powerful system to address this question. Here we describe that C/EBP with a single arginine mutation (C/EBPR35A) induces a dramatically accelerated BMT in mouse and human cells. Changes in the expression of lineage-restricted genes occur as early as within 1 hour compared to 18 hours with the wild type. Mechanistically C/EBPR35A exhibits an increased affinity for PU.1, a bi-lineage transcription factor required for C/EBP-induced BMT. The complex induces more rapid chromatin accessibility changes and an enhanced relocation (stealing) of PU.1 from B cell to myeloid gene regulatory elements. Arginine 35 is methylated by Carm1 and inhibition of the enzyme accelerates BMT, similar to the mutant. Our data suggest that the relative proportions of methylated and unmethylated C/EBP in a bipotent progenitor can determine the velocity of cell fate choice and lineage directionality.', 'published': '10.7554/elife.83951', 'node': 42510, 'link_page': 'https://www.biorxiv.org/content/10.1101/2022.10.03.510647v2?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2022.10.03.510647v2.full.pdf'}\n",
      "{'doi': '10.1101/2022.10.03.510647', 'title': 'Carm1 regulates the speed of C/EBPa-induced transdifferentiation by a cofactor stealing mechanism', 'authors': 'Garcia, G. T.; Kowenz-Leutz, E.; Tian, T. V.; Klonizakis, A.; Lerner, J.; De Andres-Aguayo, L.; Berenguer, C.; Carmona, M. P.; Casadesus, M. V.; Bulteau, R.; Francesconi, M.; Peiro, S.; Zaret, K. S.; Leutz, A.; Graf, T.', 'author_corresponding': 'Achim Leutz', 'author_corresponding_institution': 'MDC, Berlin', 'date': '2022-10-09', 'version': '3', 'license': 'cc_by_nd', 'category': 'Cell Biology', 'jatsxml': 'https://www.biorxiv.org/content/early/2022/10/09/2022.10.03.510647.source.xml', 'abstract': 'Cell fate decisions are driven by lineage-restricted transcription factors but how they are regulated is incompletely understood. The C/EBP-induced B cell to macrophage transdifferentiation (BMT) is a powerful system to address this question. Here we describe that C/EBP with a single arginine mutation (C/EBPR35A) induces a dramatically accelerated BMT in mouse and human cells. Changes in the expression of lineage-restricted genes occur as early as within 1 hour compared to 18 hours with the wild type. Mechanistically C/EBPR35A exhibits an increased affinity for PU.1, a bi-lineage transcription factor required for C/EBP-induced BMT. The complex induces more rapid chromatin accessibility changes and an enhanced relocation (stealing) of PU.1 from B cell to myeloid gene regulatory elements. Arginine 35 is methylated by Carm1 and inhibition of the enzyme accelerates BMT, similar to the mutant. Our data suggest that the relative proportions of methylated and unmethylated C/EBP in a bipotent progenitor can determine the velocity of cell fate choice and lineage directionality.', 'published': '10.7554/elife.83951', 'node': 42511, 'link_page': 'https://www.biorxiv.org/content/10.1101/2022.10.03.510647v3?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2022.10.03.510647v3.full.pdf'}\n",
      "{'doi': '10.1101/2022.10.03.510647', 'title': 'Carm1 regulates the speed of C/EBPa-induced transdifferentiation by a cofactor stealing mechanism', 'authors': 'Garcia, G. T.; Kowenz-Leutz, E.; Tian, T. V.; Klonizakis, A.; Lerner, J.; De Andres-Aguayo, L.; Berenguer, C.; Carmona, M. P.; Casadesus, M. V.; Bulteau, R.; Francesconi, M.; Leutz, A.; Zaret, K. S.; Zaret, K. S.; Peiro, S.', 'author_corresponding': 'Achim Leutz', 'author_corresponding_institution': 'MDC, Berlin', 'date': '2022-10-04', 'version': '1', 'license': 'cc_by_nc_nd', 'category': 'Cell Biology', 'jatsxml': 'https://www.biorxiv.org/content/early/2022/10/04/2022.10.03.510647.source.xml', 'abstract': 'Cell fate decisions are driven by lineage-restricted transcription factors but how they are regulated is incompletely understood. The C/EBP-induced B cell to macrophage transdifferentiation (BMT) is a powerful system to address this question. Here we describe that C/EBP with a single arginine mutation (C/EBPR35A) induces a dramatically accelerated BMT in mouse and human cells. Changes in the expression of lineage-restricted genes occur as early as within 1 hour compared to 18 hours with the wild type. Mechanistically C/EBPR35A exhibits an increased affinity for PU.1, a bi-lineage transcription factor required for C/EBP-induced BMT. The complex induces more rapid chromatin accessibility changes and an enhanced relocation (stealing) of PU.1 from B cell to myeloid gene regulatory elements. Arginine 35 is methylated by Carm1 and inhibition of the enzyme accelerates BMT, similar to the mutant. Our data suggest that the relative proportions of methylated and unmethylated C/EBP in a bipotent progenitor can determine the velocity of cell fate choice and lineage directionality.', 'published': '10.7554/elife.83951', 'node': 42509, 'link_page': 'https://www.biorxiv.org/content/10.1101/2022.10.03.510647v1?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2022.10.03.510647v1.full.pdf'}\n",
      "{'doi': '10.1101/2022.10.03.510647', 'title': 'Carm1 regulates the speed of C/EBPa-induced transdifferentiation by a cofactor stealing mechanism', 'authors': 'Garcia, G. T.; Kowenz-Leutz, E.; Tian, T. V.; Klonizakis, A.; Lerner, J.; De Andres-Aguayo, L.; Berenguer, C.; Carmona, M. P.; Casadesus, M. V.; Bulteau, R.; Francesconi, M.; Peiro, S.; Zaret, K. S.; Leutz, A.; Graf, T.', 'author_corresponding': 'Achim Leutz', 'author_corresponding_institution': 'MDC, Berlin', 'date': '2022-10-04', 'version': '2', 'license': 'cc_by_nc_nd', 'category': 'Cell Biology', 'jatsxml': 'https://www.biorxiv.org/content/early/2022/10/04/2022.10.03.510647.source.xml', 'abstract': 'Cell fate decisions are driven by lineage-restricted transcription factors but how they are regulated is incompletely understood. The C/EBP-induced B cell to macrophage transdifferentiation (BMT) is a powerful system to address this question. Here we describe that C/EBP with a single arginine mutation (C/EBPR35A) induces a dramatically accelerated BMT in mouse and human cells. Changes in the expression of lineage-restricted genes occur as early as within 1 hour compared to 18 hours with the wild type. Mechanistically C/EBPR35A exhibits an increased affinity for PU.1, a bi-lineage transcription factor required for C/EBP-induced BMT. The complex induces more rapid chromatin accessibility changes and an enhanced relocation (stealing) of PU.1 from B cell to myeloid gene regulatory elements. Arginine 35 is methylated by Carm1 and inhibition of the enzyme accelerates BMT, similar to the mutant. Our data suggest that the relative proportions of methylated and unmethylated C/EBP in a bipotent progenitor can determine the velocity of cell fate choice and lineage directionality.', 'published': '10.7554/elife.83951', 'node': 42510, 'link_page': 'https://www.biorxiv.org/content/10.1101/2022.10.03.510647v2?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2022.10.03.510647v2.full.pdf'}\n",
      "{'doi': '10.1101/2022.10.03.510647', 'title': 'Carm1 regulates the speed of C/EBPa-induced transdifferentiation by a cofactor stealing mechanism', 'authors': 'Garcia, G. T.; Kowenz-Leutz, E.; Tian, T. V.; Klonizakis, A.; Lerner, J.; De Andres-Aguayo, L.; Berenguer, C.; Carmona, M. P.; Casadesus, M. V.; Bulteau, R.; Francesconi, M.; Peiro, S.; Zaret, K. S.; Leutz, A.; Graf, T.', 'author_corresponding': 'Achim Leutz', 'author_corresponding_institution': 'MDC, Berlin', 'date': '2022-10-09', 'version': '3', 'license': 'cc_by_nd', 'category': 'Cell Biology', 'jatsxml': 'https://www.biorxiv.org/content/early/2022/10/09/2022.10.03.510647.source.xml', 'abstract': 'Cell fate decisions are driven by lineage-restricted transcription factors but how they are regulated is incompletely understood. The C/EBP-induced B cell to macrophage transdifferentiation (BMT) is a powerful system to address this question. Here we describe that C/EBP with a single arginine mutation (C/EBPR35A) induces a dramatically accelerated BMT in mouse and human cells. Changes in the expression of lineage-restricted genes occur as early as within 1 hour compared to 18 hours with the wild type. Mechanistically C/EBPR35A exhibits an increased affinity for PU.1, a bi-lineage transcription factor required for C/EBP-induced BMT. The complex induces more rapid chromatin accessibility changes and an enhanced relocation (stealing) of PU.1 from B cell to myeloid gene regulatory elements. Arginine 35 is methylated by Carm1 and inhibition of the enzyme accelerates BMT, similar to the mutant. Our data suggest that the relative proportions of methylated and unmethylated C/EBP in a bipotent progenitor can determine the velocity of cell fate choice and lineage directionality.', 'published': '10.7554/elife.83951', 'node': 42511, 'link_page': 'https://www.biorxiv.org/content/10.1101/2022.10.03.510647v3?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2022.10.03.510647v3.full.pdf'}\n",
      "-------------------------\n",
      "CPU times: user 2.5 ms, sys: 0 ns, total: 2.5 ms\n",
      "Wall time: 5.85 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "query = {\n",
    "    \"match_phrase\": {\n",
    "        \"title\": \"Carm1 regulates the speed of\"\n",
    "    }\n",
    "}\n",
    "\n",
    "# Execute the search\n",
    "results = search_data(es, index_name, query)\n",
    "\n",
    "print(\"Results: \",len(results), \"\\n\")\n",
    "# Display the results directly\n",
    "for result in results:\n",
    "    # Assuming you want to display each document's key-value pairs\n",
    "    print(result)\n",
    "\n",
    "print(\"-\" * 25)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f79a7a8-0a57-4839-9506-6e7cff75f0d5",
   "metadata": {},
   "source": [
    "\n",
    "### Multi-Match Query\n",
    "Search for a text across multiple fields.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "112d1a8a-3e4a-4246-9b71-81496ede2263",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Results:  10 \n",
      "\n",
      "{'doi': '10.1101/2021.04.02.438179', 'title': 'Mechanism of transcription modulation by the transcription-repair coupling factor', 'authors': 'Paudel, B. P.; Xu, Z.-Q.; Jergic, S.; Oakley, A. J.; Sharma, N.; Brown, S. H.; Bouwer, J. C.; Lewis, P.; Dixon, N. E.; Van Oijen, A. M.; Ghodke, H.', 'author_corresponding': 'Harshad  Ghodke', 'author_corresponding_institution': 'University of Wollongong', 'date': '2022-02-22', 'version': '3', 'license': 'cc_no', 'category': 'Biophysics', 'jatsxml': 'https://www.biorxiv.org/content/early/2022/02/22/2021.04.02.438179.source.xml', 'abstract': 'Elongation by RNA polymerase is dynamically modulated by accessory factors. The transcription-repair coupling factor (TRCF) recognizes distressed RNAPs and either rescues transcription or initiates transcription termination. Precisely how TRCFs choose to execute either outcome remains unclear. With Escherichia coli as a model, we used single-molecule assays to study dynamic modulation of elongation by Mfd, the bacterial TRCF. We found that nucleotide-bound Mfd converts the elongation complex (EC) into a catalytically poised state, presenting the EC with an opportunity to restart transcription. After long-lived residence in this catalytically poised state, ATP hydrolysis by Mfd remodels the EC through an irreversible process leading to loss of the RNA transcript. Further, biophysical studies revealed that the motor domain of Mfd binds and partially melts DNA containing a template strand overhang. The results explain pathway choice determining the fate of the EC and provide a molecular mechanism for transcription modulation by TRCF.', 'published': '10.1093/nar/gkac449', 'node': 1939, 'link_page': 'https://www.biorxiv.org/content/10.1101/2021.04.02.438179v3?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2021.04.02.438179v3.full.pdf'}\n",
      "{'doi': '10.1101/2022.11.21.517317', 'title': 'Targeted disruption of transcription bodies causes widespread activation of transcription', 'authors': 'Ugolini, M.; Kuznetsova, K.; Oda, H.; Kimura, H.; Vastenhouw, N. L.', 'author_corresponding': 'Nadine L. Vastenhouw', 'author_corresponding_institution': 'MPI-CBG, UNIL', 'date': '2022-11-21', 'version': '1', 'license': 'cc_by_nc_nd', 'category': 'Cell Biology', 'jatsxml': 'https://www.biorxiv.org/content/early/2022/11/21/2022.11.21.517317.source.xml', 'abstract': 'The localization of transcriptional activity in specialized transcription bodies is a hallmark of gene expression in eukaryotic cells. It remains unclear, however, if and how they affect gene expression. Here, we disrupted the formation of two prominent endogenous transcription bodies that mark the onset of zygotic transcription in zebrafish embryos and analysed the effect on gene expression using enriched SLAM-Seq and live-cell imaging. We find that the disruption of transcription bodies results in downregulation of hundreds of genes, providing experimental support for a model in which transcription bodies increase the efficiency of transcription. We also find that a significant number of genes are upregulated, counter to the suggested stimulatory effect of transcription bodies. These upregulated genes have accessible chromatin and are poised to be transcribed in the presence of the two transcription bodies, but they do not go into elongation. Live-cell imaging shows that the disruption of the two large transcription bodies enables these poised genes to be transcribed in ectopic transcription bodies, suggesting that the large transcription bodies sequester a pause release factor. Supporting this hypothesis, we find that CDK9, the kinase that releases paused polymerase II, is highly enriched in the two large transcription bodies. Importantly, overexpression of CDK9 in wild type embryos results in the formation of ectopic transcription bodies and thus phenocopies the removal of the two large transcription bodies. Taken together, our results show that transcription bodies regulate transcription genome-wide: the accumulation of transcriptional machinery creates a favourable environment for transcription locally, while depriving genes elsewhere in the nucleus from the same machinery.', 'published': 'NA', 'node': 49272, 'link_page': 'https://www.biorxiv.org/content/10.1101/2022.11.21.517317v1?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2022.11.21.517317v1.full.pdf'}\n",
      "{'doi': '10.1101/2022.11.21.517317', 'title': 'Targeted disruption of transcription bodies causes widespread activation of transcription', 'authors': 'Ugolini, M.; Kuznetsova, K.; Oda, H.; Kimura, H.; Vastenhouw, N. L.', 'author_corresponding': 'Nadine L. Vastenhouw', 'author_corresponding_institution': 'MPI-CBG, UNIL', 'date': '2022-11-21', 'version': '1', 'license': 'cc_by_nc_nd', 'category': 'Cell Biology', 'jatsxml': 'https://www.biorxiv.org/content/early/2022/11/21/2022.11.21.517317.source.xml', 'abstract': 'The localization of transcriptional activity in specialized transcription bodies is a hallmark of gene expression in eukaryotic cells. It remains unclear, however, if and how they affect gene expression. Here, we disrupted the formation of two prominent endogenous transcription bodies that mark the onset of zygotic transcription in zebrafish embryos and analysed the effect on gene expression using enriched SLAM-Seq and live-cell imaging. We find that the disruption of transcription bodies results in downregulation of hundreds of genes, providing experimental support for a model in which transcription bodies increase the efficiency of transcription. We also find that a significant number of genes are upregulated, counter to the suggested stimulatory effect of transcription bodies. These upregulated genes have accessible chromatin and are poised to be transcribed in the presence of the two transcription bodies, but they do not go into elongation. Live-cell imaging shows that the disruption of the two large transcription bodies enables these poised genes to be transcribed in ectopic transcription bodies, suggesting that the large transcription bodies sequester a pause release factor. Supporting this hypothesis, we find that CDK9, the kinase that releases paused polymerase II, is highly enriched in the two large transcription bodies. Importantly, overexpression of CDK9 in wild type embryos results in the formation of ectopic transcription bodies and thus phenocopies the removal of the two large transcription bodies. Taken together, our results show that transcription bodies regulate transcription genome-wide: the accumulation of transcriptional machinery creates a favourable environment for transcription locally, while depriving genes elsewhere in the nucleus from the same machinery.', 'published': 'NA', 'node': 49272, 'link_page': 'https://www.biorxiv.org/content/10.1101/2022.11.21.517317v1?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2022.11.21.517317v1.full.pdf'}\n",
      "{'doi': '10.1101/2021.04.02.438179', 'title': 'Mechanism of transcription modulation by the transcription-repair coupling factor', 'authors': 'Paudel, B. P.; Xu, Z.-Q.; Jergic, S.; Oakley, A. J.; Sharma, N.; Brown, S. H.; Bouwer, J. C.; Lewis, P.; Dixon, N. E.; Van Oijen, A. M.; Ghodke, H.', 'author_corresponding': 'Harshad  Ghodke', 'author_corresponding_institution': 'University of Wollongong', 'date': '2022-02-22', 'version': '3', 'license': 'cc_no', 'category': 'Biophysics', 'jatsxml': 'https://www.biorxiv.org/content/early/2022/02/22/2021.04.02.438179.source.xml', 'abstract': 'Elongation by RNA polymerase is dynamically modulated by accessory factors. The transcription-repair coupling factor (TRCF) recognizes distressed RNAPs and either rescues transcription or initiates transcription termination. Precisely how TRCFs choose to execute either outcome remains unclear. With Escherichia coli as a model, we used single-molecule assays to study dynamic modulation of elongation by Mfd, the bacterial TRCF. We found that nucleotide-bound Mfd converts the elongation complex (EC) into a catalytically poised state, presenting the EC with an opportunity to restart transcription. After long-lived residence in this catalytically poised state, ATP hydrolysis by Mfd remodels the EC through an irreversible process leading to loss of the RNA transcript. Further, biophysical studies revealed that the motor domain of Mfd binds and partially melts DNA containing a template strand overhang. The results explain pathway choice determining the fate of the EC and provide a molecular mechanism for transcription modulation by TRCF.', 'published': '10.1093/nar/gkac449', 'node': 1939, 'link_page': 'https://www.biorxiv.org/content/10.1101/2021.04.02.438179v3?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2021.04.02.438179v3.full.pdf'}\n",
      "{'doi': '10.1101/2022.06.13.495463', 'title': 'Nanog organizes transcription bodies', 'authors': 'Kuznetsova, K.; Ugolini, M.; Wu, E.; Lalit, M.; Oda, H.; Sato, Y.; Kimura, H.; Jug, F.; Vastenhouw, N.', 'author_corresponding': 'Nadine  Vastenhouw', 'author_corresponding_institution': 'Max Planck Institute of Molecular Cell Biology and Genetics, Dresden, Germany; Center for Integrative Genomics, University of Lausanne, Lausanne, Switzerland', 'date': '2022-06-13', 'version': '1', 'license': 'cc_by_nc_nd', 'category': 'Cell Biology', 'jatsxml': 'https://www.biorxiv.org/content/early/2022/06/13/2022.06.13.495463.source.xml', 'abstract': 'The localization of transcriptional activity in specialized transcription bodies is a hallmark of gene expression in eukaryotic cells. How proteins of the transcriptional machinery come together to form such bodies, however, is unclear. Here, we take advantage of two large, isolated, and long-lived transcription bodies that reproducibly form during early zebrafish embryogenesis, to characterize the dynamics of transcription body formation. Once formed, these transcription bodies are enriched for initiating and elongating RNA polymerase II, as well as the transcription factors Nanog and Sox19b. Analyzing the events leading up to transcription, we find that Nanog and Sox19b cluster prior to transcription, and independently of RNA accumulation. The clustering of transcription factors is sequential; Nanog clusters first, and this is required for the clustering of Sox19b and the initiation of transcription. Mutant analysis revealed that both the DNA-binding domain, as well as one of the two intrinsically disordered regions of Nanog are required to organize the two bodies of transcriptional activity. Taken together, our data suggests that the clustering of transcription factors dictates the formation of transcription bodies.\\n\\nHIGHLIGHTSO_LITranscription factors cluster prior to, and independently of transcription\\nC_LIO_LINanog organizes transcription bodies: it is required for the clustering of Sox19b as well as RNA polymerase II\\nC_LIO_LIThis organizing activity requires its DNA binding domain as well as one of its intrinsically disordered regions\\nC_LIO_LITranscription elongation results in the disassembly of transcription factor clusters\\nC_LI', 'published': '10.1016/j.cub.2022.11.015', 'node': 27656, 'link_page': 'https://www.biorxiv.org/content/10.1101/2022.06.13.495463v1?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2022.06.13.495463v1.full.pdf'}\n",
      "{'doi': '10.1101/2022.06.13.495463', 'title': 'Nanog organizes transcription bodies', 'authors': 'Kuznetsova, K.; Ugolini, M.; Wu, E.; Lalit, M.; Oda, H.; Sato, Y.; Kimura, H.; Jug, F.; Vastenhouw, N.', 'author_corresponding': 'Nadine  Vastenhouw', 'author_corresponding_institution': 'Max Planck Institute of Molecular Cell Biology and Genetics, Dresden, Germany; Center for Integrative Genomics, University of Lausanne, Lausanne, Switzerland', 'date': '2022-06-13', 'version': '1', 'license': 'cc_by_nc_nd', 'category': 'Cell Biology', 'jatsxml': 'https://www.biorxiv.org/content/early/2022/06/13/2022.06.13.495463.source.xml', 'abstract': 'The localization of transcriptional activity in specialized transcription bodies is a hallmark of gene expression in eukaryotic cells. How proteins of the transcriptional machinery come together to form such bodies, however, is unclear. Here, we take advantage of two large, isolated, and long-lived transcription bodies that reproducibly form during early zebrafish embryogenesis, to characterize the dynamics of transcription body formation. Once formed, these transcription bodies are enriched for initiating and elongating RNA polymerase II, as well as the transcription factors Nanog and Sox19b. Analyzing the events leading up to transcription, we find that Nanog and Sox19b cluster prior to transcription, and independently of RNA accumulation. The clustering of transcription factors is sequential; Nanog clusters first, and this is required for the clustering of Sox19b and the initiation of transcription. Mutant analysis revealed that both the DNA-binding domain, as well as one of the two intrinsically disordered regions of Nanog are required to organize the two bodies of transcriptional activity. Taken together, our data suggests that the clustering of transcription factors dictates the formation of transcription bodies.\\n\\nHIGHLIGHTSO_LITranscription factors cluster prior to, and independently of transcription\\nC_LIO_LINanog organizes transcription bodies: it is required for the clustering of Sox19b as well as RNA polymerase II\\nC_LIO_LIThis organizing activity requires its DNA binding domain as well as one of its intrinsically disordered regions\\nC_LIO_LITranscription elongation results in the disassembly of transcription factor clusters\\nC_LI', 'published': '10.1016/j.cub.2022.11.015', 'node': 27656, 'link_page': 'https://www.biorxiv.org/content/10.1101/2022.06.13.495463v1?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2022.06.13.495463v1.full.pdf'}\n",
      "{'doi': '10.1101/2021.02.17.431652', 'title': 'How the replication and transcription complex functions in jumping transcription of SARS-CoV-2', 'authors': 'Liang, J.; Shi, J.; Chen, S.; Duan, G.; Yang, F.; Zhi, C.; Li, X.; Ruan, J.; Mi, D.; Shan, G.', 'author_corresponding': 'Gao  Shan', 'author_corresponding_institution': 'Nankai University', 'date': '2022-04-12', 'version': '2', 'license': 'cc_by_nc', 'category': 'Bioinformatics', 'jatsxml': 'https://www.biorxiv.org/content/early/2022/04/12/2021.02.17.431652.source.xml', 'abstract': 'BackgroundCoronavirus disease 2019 (COVID-19) is caused by severe acute respiratory syndrome coronavirus 2 (SARS-CoV-2). Although unprecedented efforts are underway to develop therapeutic strategies against this disease, scientists have acquired only a little knowledge regarding the structures and functions of the CoV replication and transcription complex (RTC) and 16 non-structural proteins, named NSP1-16.\\n\\nResultsIn the present study, we proposed a two-route model to answer how the RTC functions in the jumping transcription of CoVs. The key step leading to this model was that the motif AAACH for METTL3 recognition flanking the transcription regulatory sequence (TRS) motif was discovered to determine the m6A methylation of SARS-CoV-2 RNAs, by reanalyzing public Nanopore RNA-seq data. As the most important finding, TRS hairpins were reported for the first time to interpret NSP15 cleavage, RNA methylation of CoVs and their association at the molecular level. In addition, we reported canonical TRS motifs of all CoVs to prove the importance of our findings.\\n\\nConclusionsThe main conclusions are: (1) TRS hairpins can be used to identify recombination regions in CoV genomes; (2) RNA methylation of CoVs participates in the determination of the RNA secondary structures by affecting the formation of base pairing; and (3) The eventual determination of the CoV RTC global structure needs to consider METTL3 in the experimental design. Our findings enrich fundamental knowledge in the field of gene expression and its regulation, providing a crucial basis for future studies.', 'published': '10.3389/fgene.2022.904513', 'node': 1602, 'link_page': 'https://www.biorxiv.org/content/10.1101/2021.02.17.431652v2?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2021.02.17.431652v2.full.pdf'}\n",
      "{'doi': '10.1101/2023.02.08.527627', 'title': 'Transcription activator WCC recruits deacetylase HDA3 to control transcription dynamics and bursting in Neurospora', 'authors': 'Oehler, M.; Diernfellner, A. C. R.; Brunner, M.', 'author_corresponding': 'Michael Brunner', 'author_corresponding_institution': 'Heidelberg University Biochemistry Center', 'date': '2023-02-08', 'version': '1', 'license': 'cc_by_nc_nd', 'category': 'Molecular Biology', 'jatsxml': 'https://www.biorxiv.org/content/early/2023/02/08/2023.02.08.527627.source.xml', 'abstract': 'RNA polymerase II initiates transcription either randomly or in bursts. We examined the light-dependent transcriptional activator White Collar Complex (WCC) of Neurospora to characterize the transcriptional dynamics of the strong vivid (vvd) promoter and the weaker frequency (frq) promoter. We show that WCC is not only an activator but also represses transcription by recruiting histone deacetylase 3 (HDA3). Our data suggest that bursts of frq transcription are governed by a long-lived refractory state established and maintained by WCC and HDA3 at the core promoter, whereas transcription of vvd is determined by WCC binding dynamics at an upstream activating sequence. Thus, in addition to stochastic binding of transcription factors, transcription factor-mediated repression may also influence transcriptional bursting.\\n\\nTEASERBalanced interaction of transcription factor with coactivator and corepressors determines transcription dynamics and bursting.', 'published': '10.1126/sciadv.adh0721', 'node': 59710, 'link_page': 'https://www.biorxiv.org/content/10.1101/2023.02.08.527627v1?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2023.02.08.527627v1.full.pdf'}\n",
      "{'doi': '10.1101/2023.02.08.527627', 'title': 'Transcription activator WCC recruits deacetylase HDA3 to control transcription dynamics and bursting in Neurospora', 'authors': 'Oehler, M.; Diernfellner, A. C. R.; Brunner, M.', 'author_corresponding': 'Michael Brunner', 'author_corresponding_institution': 'Heidelberg University Biochemistry Center', 'date': '2023-02-08', 'version': '1', 'license': 'cc_by_nc_nd', 'category': 'Molecular Biology', 'jatsxml': 'https://www.biorxiv.org/content/early/2023/02/08/2023.02.08.527627.source.xml', 'abstract': 'RNA polymerase II initiates transcription either randomly or in bursts. We examined the light-dependent transcriptional activator White Collar Complex (WCC) of Neurospora to characterize the transcriptional dynamics of the strong vivid (vvd) promoter and the weaker frequency (frq) promoter. We show that WCC is not only an activator but also represses transcription by recruiting histone deacetylase 3 (HDA3). Our data suggest that bursts of frq transcription are governed by a long-lived refractory state established and maintained by WCC and HDA3 at the core promoter, whereas transcription of vvd is determined by WCC binding dynamics at an upstream activating sequence. Thus, in addition to stochastic binding of transcription factors, transcription factor-mediated repression may also influence transcriptional bursting.\\n\\nTEASERBalanced interaction of transcription factor with coactivator and corepressors determines transcription dynamics and bursting.', 'published': '10.1126/sciadv.adh0721', 'node': 59710, 'link_page': 'https://www.biorxiv.org/content/10.1101/2023.02.08.527627v1?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2023.02.08.527627v1.full.pdf'}\n",
      "{'doi': '10.1101/2021.02.17.431652', 'title': 'How the replication and transcription complex functions in jumping transcription of SARS-CoV-2', 'authors': 'Liang, J.; Shi, J.; Chen, S.; Duan, G.; Yang, F.; Zhi, C.; Li, X.; Ruan, J.; Mi, D.; Shan, G.', 'author_corresponding': 'Gao  Shan', 'author_corresponding_institution': 'Nankai University', 'date': '2022-04-12', 'version': '2', 'license': 'cc_by_nc', 'category': 'Bioinformatics', 'jatsxml': 'https://www.biorxiv.org/content/early/2022/04/12/2021.02.17.431652.source.xml', 'abstract': 'BackgroundCoronavirus disease 2019 (COVID-19) is caused by severe acute respiratory syndrome coronavirus 2 (SARS-CoV-2). Although unprecedented efforts are underway to develop therapeutic strategies against this disease, scientists have acquired only a little knowledge regarding the structures and functions of the CoV replication and transcription complex (RTC) and 16 non-structural proteins, named NSP1-16.\\n\\nResultsIn the present study, we proposed a two-route model to answer how the RTC functions in the jumping transcription of CoVs. The key step leading to this model was that the motif AAACH for METTL3 recognition flanking the transcription regulatory sequence (TRS) motif was discovered to determine the m6A methylation of SARS-CoV-2 RNAs, by reanalyzing public Nanopore RNA-seq data. As the most important finding, TRS hairpins were reported for the first time to interpret NSP15 cleavage, RNA methylation of CoVs and their association at the molecular level. In addition, we reported canonical TRS motifs of all CoVs to prove the importance of our findings.\\n\\nConclusionsThe main conclusions are: (1) TRS hairpins can be used to identify recombination regions in CoV genomes; (2) RNA methylation of CoVs participates in the determination of the RNA secondary structures by affecting the formation of base pairing; and (3) The eventual determination of the CoV RTC global structure needs to consider METTL3 in the experimental design. Our findings enrich fundamental knowledge in the field of gene expression and its regulation, providing a crucial basis for future studies.', 'published': '10.3389/fgene.2022.904513', 'node': 1602, 'link_page': 'https://www.biorxiv.org/content/10.1101/2021.02.17.431652v2?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2021.02.17.431652v2.full.pdf'}\n",
      "-------------------------\n",
      "CPU times: user 3.92 ms, sys: 296 µs, total: 4.21 ms\n",
      "Wall time: 10.1 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "query = {\n",
    "    \"multi_match\": {\n",
    "        \"query\": \"transcription\",\n",
    "        \"fields\": [\"title\", \"abstract\"]\n",
    "    }\n",
    "}\n",
    "\n",
    "# Execute the search\n",
    "results = search_data(es, index_name, query)\n",
    "\n",
    "print(\"Results: \",len(results), \"\\n\")\n",
    "# Display the results directly\n",
    "for result in results:\n",
    "    # Assuming you want to display each document's key-value pairs\n",
    "    print(result)\n",
    "\n",
    "print(\"-\" * 25)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1189211e-fd75-4f5d-91a0-31f5224e7c5f",
   "metadata": {},
   "source": [
    "### Advanced Filtering: Beyond Keywords"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a62554d3-fd0f-4dff-bb59-98a63c2e711e",
   "metadata": {},
   "source": [
    "Filtering allows for more refined searches, such as retrieving documents within a specific date range or by particular authors, enhancing the precision of your data analysis.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "6f3125f7-3853-439a-b003-b5e939b5e809",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Results:  10 \n",
      "\n",
      "{'doi': '10.1101/2023.05.24.542151', 'title': 'POMBOX: a fission yeast toolkit for molecular and synthetic biology', 'authors': 'Hebra, T.; Smrckova, H.; Elkatmis, B.; Prevorovsky, M.; Pluskal, T.', 'author_corresponding': 'Tomas Pluskal', 'author_corresponding_institution': 'Institute of Organic Chemistry and Biochemistry of the Czech Academy of Sciences, Praha, Czech Republic', 'date': '2023-05-24', 'version': '1', 'license': 'cc_by_nc_nd', 'category': 'Synthetic Biology', 'jatsxml': 'https://www.biorxiv.org/content/early/2023/05/24/2023.05.24.542151.source.xml', 'abstract': 'Schizosaccharomyces pombe is a popular model organism in molecular biology and cell physiology. With its ease of genetic manipulation and growth, supported by in-depth functional annotation in the PomBase database and genome-wide metabolic models, S. pombe is an attractive option for synthetic biology applications. However, S. pombe currently lacks modular tools for generating genetic circuits with more than one transcriptional unit. We have developed a toolkit to address this issue. Adapted from the MoClo- YTK plasmid kit for Saccharomyces cerevisiae and using the same Golden Gate grammar, our POMBOX toolkit is designed to facilitate the fast, efficient and modular construction of genetic circuits in S. pombe. It allows for interoperability when working with DNA sequences that are functional in both S. cerevisiae and S. pombe (e.g. protein tag, antibiotic resistance cassette, coding sequences). Moreover, POMBOX enables the modular assembly of multi-gene pathways and increases possible pathway length from 6 to 12 transcriptional units. We also adapted the stable integration vector homology arms to Golden Gate assembly and tested the genomic integration success rate depending on different sequence sizes, from four to twenty-four kilobases. We included fourteen S. pombe promoters that we characterized for two fluorescent proteins, in both minimal defined media (EMM2) and complex media (YES). Then we tested six S. cerevisiae and six synthetic terminators in S. pombe. Finally, we used the POMBOX kit for a synthetic biology application in metabolic engineering and expressed plant enzymes in S. pombe to produce specialized metabolite precursors, namely methylxanthine, amorpha-4,11-diene and cinnamic acid from the purine, mevalonate and amino acid pathways.', 'published': 'NA', 'node': 74136, 'link_page': 'https://www.biorxiv.org/content/10.1101/2023.05.24.542151v1?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2023.05.24.542151v1.full.pdf'}\n",
      "{'doi': '10.1101/2023.05.24.542151', 'title': 'POMBOX: a fission yeast toolkit for molecular and synthetic biology', 'authors': 'Hebra, T.; Smrckova, H.; Elkatmis, B.; Prevorovsky, M.; Pluskal, T.', 'author_corresponding': 'Tomas Pluskal', 'author_corresponding_institution': 'Institute of Organic Chemistry and Biochemistry of the Czech Academy of Sciences, Praha, Czech Republic', 'date': '2023-05-24', 'version': '1', 'license': 'cc_by_nc_nd', 'category': 'Synthetic Biology', 'jatsxml': 'https://www.biorxiv.org/content/early/2023/05/24/2023.05.24.542151.source.xml', 'abstract': 'Schizosaccharomyces pombe is a popular model organism in molecular biology and cell physiology. With its ease of genetic manipulation and growth, supported by in-depth functional annotation in the PomBase database and genome-wide metabolic models, S. pombe is an attractive option for synthetic biology applications. However, S. pombe currently lacks modular tools for generating genetic circuits with more than one transcriptional unit. We have developed a toolkit to address this issue. Adapted from the MoClo- YTK plasmid kit for Saccharomyces cerevisiae and using the same Golden Gate grammar, our POMBOX toolkit is designed to facilitate the fast, efficient and modular construction of genetic circuits in S. pombe. It allows for interoperability when working with DNA sequences that are functional in both S. cerevisiae and S. pombe (e.g. protein tag, antibiotic resistance cassette, coding sequences). Moreover, POMBOX enables the modular assembly of multi-gene pathways and increases possible pathway length from 6 to 12 transcriptional units. We also adapted the stable integration vector homology arms to Golden Gate assembly and tested the genomic integration success rate depending on different sequence sizes, from four to twenty-four kilobases. We included fourteen S. pombe promoters that we characterized for two fluorescent proteins, in both minimal defined media (EMM2) and complex media (YES). Then we tested six S. cerevisiae and six synthetic terminators in S. pombe. Finally, we used the POMBOX kit for a synthetic biology application in metabolic engineering and expressed plant enzymes in S. pombe to produce specialized metabolite precursors, namely methylxanthine, amorpha-4,11-diene and cinnamic acid from the purine, mevalonate and amino acid pathways.', 'published': 'NA', 'node': 74136, 'link_page': 'https://www.biorxiv.org/content/10.1101/2023.05.24.542151v1?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2023.05.24.542151v1.full.pdf'}\n",
      "{'doi': '10.1101/2022.06.30.498253', 'title': 'A DARPin-based molecular toolset to probe gephyrin and inhibitory synapse biology', 'authors': 'Campbell, B. F.; Dittmann, A.; Dreier, B.; Plückthun, A.; Tyagarajan, S. K.', 'author_corresponding': 'Shiva K  Tyagarajan', 'author_corresponding_institution': 'University of Zurich', 'date': '2022-07-03', 'version': '1', 'license': 'cc_by', 'category': 'Neuroscience', 'jatsxml': 'https://www.biorxiv.org/content/early/2022/07/03/2022.06.30.498253.source.xml', 'abstract': 'Neuroscience currently requires the use of antibodies to study synaptic proteins, where antibody binding is used as a correlate to define the presence, plasticity, and regulation of synapses. Gephyrin is an inhibitory synaptic scaffolding protein used to mark GABAergic and glycinergic postsynaptic sites. Despite the importance of gephyrin in modulating inhibitory transmission, its study is currently limited by the tractability of available reagents. Designed Ankyrin Repeat Proteins (DARPins) are a class of synthetic protein binder derived from diverse libraries by in vitro selection, and tested by high-throughput screening to produce specific binders. In order to generate a functionally diverse toolset for studying inhibitory synapses, we screened a DARPin library against gephyrin mutants representing both phosphorylated and dephosphorylated states. We validated the robust use of anti-gephyrin DARPin clones for morphological identification of gephyrin clusters in rodent neuron culture and brain tissue, discovering previously overlooked clusters. This DARPin-based toolset includes clones with heterogenous gephyrin binding modes that allowed for identification of the most extensive gephyrin interactome to date, and defined novel classes of putative interactors, creating a framework for understanding gephyrins non-synaptic functions. This study demonstrates anti-gephyrin DARPins as a versatile platform for studying inhibitory synapses in an unprecedented manner.', 'published': '10.7554/eLife.80895', 'node': 30322, 'link_page': 'https://www.biorxiv.org/content/10.1101/2022.06.30.498253v1?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2022.06.30.498253v1.full.pdf'}\n",
      "{'doi': '10.1101/2023.08.11.553013', 'title': 'Efficient Natural Plasmid Transformation of Vibrio natriegens Enables Zero-capital Molecular Biology', 'authors': 'Specht, D. A.; Sheppard, T. J.; Kennedy, F.; Li, S.; Gadikota, G.; Barstow, B.', 'author_corresponding': 'Buz Barstow', 'author_corresponding_institution': 'Cornell University', 'date': '2023-08-13', 'version': '1', 'license': 'cc_no', 'category': 'Synthetic Biology', 'jatsxml': 'https://www.biorxiv.org/content/early/2023/08/13/2023.08.11.553013.source.xml', 'abstract': 'The fast-growing microbe Vibrio natriegens is capable of natural transformation where it draws DNA in from media via an active process under physiological conditions. Using an engineered strain with a genomic copy of the master competence regulator tfoX from Vibrio cholera in combination with a new minimal competence media (MCM) that uses acetate as an energy source, we demonstrate naturally competent cells which are created, transformed, and recovered entirely in the same media, without exchange or addition of new media. Cells are naturally competent to plasmids, recombination with linear DNA, and co-transformation of both to select for scarless and markerless genomic edits. The entire process is simple and inexpensive, requiring no capital equipment for an entirely room temperature process (Zero Capital protocol, 104 cfu/{micro}g), or just an incubator (High Efficiency protocol, 105-6 cfu/{micro}g). These cells retain their naturally competent state when frozen and are transformable immediately upon thawing like a typical chemical or electrochemical competent cell. Since the optimized transformation protocol requires only 50 minutes of hands-on time, and V. natriegens grows quickly even on plates, a transformation started at 9 AM yields abundant culturable single colonies by 5 PM. Further, because all stages of transformation occur in the same media, and the process can be arbitrarily scaled in volume, this natural competence strain and media could be ideal for automated directed evolution applications. As a result, naturally competent V. natriegens could compete with E. coli as an excellent chassis for low-cost and highly scalable synthetic biology.', 'published': 'NA', 'node': 84610, 'link_page': 'https://www.biorxiv.org/content/10.1101/2023.08.11.553013v1?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2023.08.11.553013v1.full.pdf'}\n",
      "{'doi': '10.1101/2022.06.30.498253', 'title': 'A DARPin-based molecular toolset to probe gephyrin and inhibitory synapse biology', 'authors': 'Campbell, B. F.; Dittmann, A.; Dreier, B.; Plückthun, A.; Tyagarajan, S. K.', 'author_corresponding': 'Shiva K  Tyagarajan', 'author_corresponding_institution': 'University of Zurich', 'date': '2022-07-03', 'version': '1', 'license': 'cc_by', 'category': 'Neuroscience', 'jatsxml': 'https://www.biorxiv.org/content/early/2022/07/03/2022.06.30.498253.source.xml', 'abstract': 'Neuroscience currently requires the use of antibodies to study synaptic proteins, where antibody binding is used as a correlate to define the presence, plasticity, and regulation of synapses. Gephyrin is an inhibitory synaptic scaffolding protein used to mark GABAergic and glycinergic postsynaptic sites. Despite the importance of gephyrin in modulating inhibitory transmission, its study is currently limited by the tractability of available reagents. Designed Ankyrin Repeat Proteins (DARPins) are a class of synthetic protein binder derived from diverse libraries by in vitro selection, and tested by high-throughput screening to produce specific binders. In order to generate a functionally diverse toolset for studying inhibitory synapses, we screened a DARPin library against gephyrin mutants representing both phosphorylated and dephosphorylated states. We validated the robust use of anti-gephyrin DARPin clones for morphological identification of gephyrin clusters in rodent neuron culture and brain tissue, discovering previously overlooked clusters. This DARPin-based toolset includes clones with heterogenous gephyrin binding modes that allowed for identification of the most extensive gephyrin interactome to date, and defined novel classes of putative interactors, creating a framework for understanding gephyrins non-synaptic functions. This study demonstrates anti-gephyrin DARPins as a versatile platform for studying inhibitory synapses in an unprecedented manner.', 'published': '10.7554/eLife.80895', 'node': 30322, 'link_page': 'https://www.biorxiv.org/content/10.1101/2022.06.30.498253v1?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2022.06.30.498253v1.full.pdf'}\n",
      "{'doi': '10.1101/2023.08.11.553013', 'title': 'Efficient Natural Plasmid Transformation of Vibrio natriegens Enables Zero-capital Molecular Biology', 'authors': 'Specht, D. A.; Sheppard, T. J.; Kennedy, F.; Li, S.; Gadikota, G.; Barstow, B.', 'author_corresponding': 'Buz Barstow', 'author_corresponding_institution': 'Cornell University', 'date': '2023-08-13', 'version': '1', 'license': 'cc_no', 'category': 'Synthetic Biology', 'jatsxml': 'https://www.biorxiv.org/content/early/2023/08/13/2023.08.11.553013.source.xml', 'abstract': 'The fast-growing microbe Vibrio natriegens is capable of natural transformation where it draws DNA in from media via an active process under physiological conditions. Using an engineered strain with a genomic copy of the master competence regulator tfoX from Vibrio cholera in combination with a new minimal competence media (MCM) that uses acetate as an energy source, we demonstrate naturally competent cells which are created, transformed, and recovered entirely in the same media, without exchange or addition of new media. Cells are naturally competent to plasmids, recombination with linear DNA, and co-transformation of both to select for scarless and markerless genomic edits. The entire process is simple and inexpensive, requiring no capital equipment for an entirely room temperature process (Zero Capital protocol, 104 cfu/{micro}g), or just an incubator (High Efficiency protocol, 105-6 cfu/{micro}g). These cells retain their naturally competent state when frozen and are transformable immediately upon thawing like a typical chemical or electrochemical competent cell. Since the optimized transformation protocol requires only 50 minutes of hands-on time, and V. natriegens grows quickly even on plates, a transformation started at 9 AM yields abundant culturable single colonies by 5 PM. Further, because all stages of transformation occur in the same media, and the process can be arbitrarily scaled in volume, this natural competence strain and media could be ideal for automated directed evolution applications. As a result, naturally competent V. natriegens could compete with E. coli as an excellent chassis for low-cost and highly scalable synthetic biology.', 'published': 'NA', 'node': 84610, 'link_page': 'https://www.biorxiv.org/content/10.1101/2023.08.11.553013v1?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2023.08.11.553013v1.full.pdf'}\n",
      "{'doi': '10.1101/2023.04.11.535671', 'title': 'Molecular profiling of high-level athlete skeletal muscle after acute exercise - a systems biology approach', 'authors': 'Reitzner, S. M.; Emanuelsson, E. B.; Arif, M.; Kaczkowski, B.; Kwon, A. T.-J.; Mardinoglu, A.; Arner, E.; Chapman, M. A.; Sundberg, C. J.', 'author_corresponding': 'Stefan Markus Reitzner', 'author_corresponding_institution': 'Department of Physiology and Pharmacology, Karolinska Institutet', 'date': '2023-04-11', 'version': '1', 'license': 'cc_by_nc_nd', 'category': 'Physiology', 'jatsxml': 'https://www.biorxiv.org/content/early/2023/04/11/2023.04.11.535671.source.xml', 'abstract': 'Life-long high-level exercise training leads to improvements in physical performance and multi-tissue adaptation following changes in molecular pathways. While skeletal muscle baseline differences between exercise-trained and untrained individuals have been previously investigated, it remains unclear how acute exercise multi-omics are influenced by training history. We recruited and extensively characterized 24 individuals categorized as endurance athletes, strength athletes or control subjects. Multi-omics profiling was performed from skeletal muscle before and at three time-points after endurance or resistance exercise sessions. Timeseries multi-omics analysis revealed distinct differences in molecular processes such as fatty- and amino acid metabolism and for transcription factors such as HIF1A and the MYF-family between both exercise history and acute form of exercise. Furthermore, we found a \"transcriptional specialization effect\" by transcriptional narrowing and intensification. Finally, we performed multi-omics network analysis and clustering, providing a novel resource of skeletal muscle transcriptomic and metabolomic profiling in highly trained and untrained individuals.', 'published': 'NA', 'node': 68337, 'link_page': 'https://www.biorxiv.org/content/10.1101/2023.04.11.535671v1?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2023.04.11.535671v1.full.pdf'}\n",
      "{'doi': '10.1101/2023.04.11.535671', 'title': 'Molecular profiling of high-level athlete skeletal muscle after acute exercise - a systems biology approach', 'authors': 'Reitzner, S. M.; Emanuelsson, E. B.; Arif, M.; Kaczkowski, B.; Kwon, A. T.-J.; Mardinoglu, A.; Arner, E.; Chapman, M. A.; Sundberg, C. J.', 'author_corresponding': 'Stefan Markus Reitzner', 'author_corresponding_institution': 'Department of Physiology and Pharmacology, Karolinska Institutet', 'date': '2023-04-11', 'version': '1', 'license': 'cc_by_nc_nd', 'category': 'Physiology', 'jatsxml': 'https://www.biorxiv.org/content/early/2023/04/11/2023.04.11.535671.source.xml', 'abstract': 'Life-long high-level exercise training leads to improvements in physical performance and multi-tissue adaptation following changes in molecular pathways. While skeletal muscle baseline differences between exercise-trained and untrained individuals have been previously investigated, it remains unclear how acute exercise multi-omics are influenced by training history. We recruited and extensively characterized 24 individuals categorized as endurance athletes, strength athletes or control subjects. Multi-omics profiling was performed from skeletal muscle before and at three time-points after endurance or resistance exercise sessions. Timeseries multi-omics analysis revealed distinct differences in molecular processes such as fatty- and amino acid metabolism and for transcription factors such as HIF1A and the MYF-family between both exercise history and acute form of exercise. Furthermore, we found a \"transcriptional specialization effect\" by transcriptional narrowing and intensification. Finally, we performed multi-omics network analysis and clustering, providing a novel resource of skeletal muscle transcriptomic and metabolomic profiling in highly trained and untrained individuals.', 'published': 'NA', 'node': 68337, 'link_page': 'https://www.biorxiv.org/content/10.1101/2023.04.11.535671v1?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2023.04.11.535671v1.full.pdf'}\n",
      "{'doi': '10.1101/2022.02.03.475821', 'title': 'Abnormal molecular signatures of inflammation, energy metabolism and vesicle biology in human Huntington disease peripheral tissues', 'authors': 'Neueder, A.; Kojer, K.; Hering, T.; Lavery, D. J.; Chen, J.; Birth, N.; Hallitsch, J.; Trautmann, S.; Parker, J.; Flower, M.; Sethi, H.; Haider, S.; Lee, J.-M.; Tabrizi, S. J.; Orth, M.', 'author_corresponding': 'Michael  Orth', 'author_corresponding_institution': 'Department of Neurology; Ulm University, 89081 Ulm, Germany; Swiss Huntington Centre; Neurozentrum, Siloah AG, Worbstr. 312, 3073 Guemligen, Switzerland; Univer', 'date': '2022-02-06', 'version': '1', 'license': 'cc_no', 'category': 'Neuroscience', 'jatsxml': 'https://www.biorxiv.org/content/early/2022/02/06/2022.02.03.475821.source.xml', 'abstract': 'BackgroundA major challenge in neurodegenerative diseases concerns identifying biological disease signatures that track with disease progression or respond to an intervention. Several clinical trials in Huntington disease (HD), an inherited, progressive neurodegenerative disease, are currently ongoing. Therefore, we examined whether peripheral tissues can serve as a source of readily accessible biological signatures at the RNA and protein level in HD patients.\\n\\nResultsWe generated large, high-quality human datasets from skeletal muscle, skin and adipose tissue to probe molecular changes in human premanifest and early manifest HD patients - those most likely involved in clinical trials. In-depth single nucleotide polymorphism data across the HTT gene will facilitate the use of the generated primary- and iPSC cell lines in allele-specific targeting approaches. The analysis of the transcriptomics and proteomics data shows robust, stage-dependent dysregulation. Gene ontology analysis confirmed the involvement of inflammation and energy metabolism in peripheral HD pathogenesis. Furthermore, we observed changes in the homeostasis of extracellular vesicles, where we found consistent changes of genes and proteins involved in this process.\\n\\nConclusionsOur  omics data document the involvement of inflammation, energy metabolism and extracellular vesicle homeostasis. This demonstrates the potential to identify biological signatures from peripheral tissues in HD suitable as biomarkers in clinical trials. Together with the primary cell lines established from peripheral tissues and a large panel of iPSC lines that can serve as human models of HD, the generated data are a valuable and unique resource to advance the current understanding of molecular mechanisms driving HD pathogenesis.', 'published': '10.1186/s13059-022-02752-5', 'node': 10990, 'link_page': 'https://www.biorxiv.org/content/10.1101/2022.02.03.475821v1?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2022.02.03.475821v1.full.pdf'}\n",
      "{'doi': '10.1101/2022.02.03.475821', 'title': 'Abnormal molecular signatures of inflammation, energy metabolism and vesicle biology in human Huntington disease peripheral tissues', 'authors': 'Neueder, A.; Kojer, K.; Hering, T.; Lavery, D. J.; Chen, J.; Birth, N.; Hallitsch, J.; Trautmann, S.; Parker, J.; Flower, M.; Sethi, H.; Haider, S.; Lee, J.-M.; Tabrizi, S. J.; Orth, M.', 'author_corresponding': 'Michael  Orth', 'author_corresponding_institution': 'Department of Neurology; Ulm University, 89081 Ulm, Germany; Swiss Huntington Centre; Neurozentrum, Siloah AG, Worbstr. 312, 3073 Guemligen, Switzerland; Univer', 'date': '2022-02-06', 'version': '1', 'license': 'cc_no', 'category': 'Neuroscience', 'jatsxml': 'https://www.biorxiv.org/content/early/2022/02/06/2022.02.03.475821.source.xml', 'abstract': 'BackgroundA major challenge in neurodegenerative diseases concerns identifying biological disease signatures that track with disease progression or respond to an intervention. Several clinical trials in Huntington disease (HD), an inherited, progressive neurodegenerative disease, are currently ongoing. Therefore, we examined whether peripheral tissues can serve as a source of readily accessible biological signatures at the RNA and protein level in HD patients.\\n\\nResultsWe generated large, high-quality human datasets from skeletal muscle, skin and adipose tissue to probe molecular changes in human premanifest and early manifest HD patients - those most likely involved in clinical trials. In-depth single nucleotide polymorphism data across the HTT gene will facilitate the use of the generated primary- and iPSC cell lines in allele-specific targeting approaches. The analysis of the transcriptomics and proteomics data shows robust, stage-dependent dysregulation. Gene ontology analysis confirmed the involvement of inflammation and energy metabolism in peripheral HD pathogenesis. Furthermore, we observed changes in the homeostasis of extracellular vesicles, where we found consistent changes of genes and proteins involved in this process.\\n\\nConclusionsOur  omics data document the involvement of inflammation, energy metabolism and extracellular vesicle homeostasis. This demonstrates the potential to identify biological signatures from peripheral tissues in HD suitable as biomarkers in clinical trials. Together with the primary cell lines established from peripheral tissues and a large panel of iPSC lines that can serve as human models of HD, the generated data are a valuable and unique resource to advance the current understanding of molecular mechanisms driving HD pathogenesis.', 'published': '10.1186/s13059-022-02752-5', 'node': 10990, 'link_page': 'https://www.biorxiv.org/content/10.1101/2022.02.03.475821v1?versioned=TRUE', 'link_pdf': 'https://www.biorxiv.org/content/10.1101/2022.02.03.475821v1.full.pdf'}\n",
      "-------------------------\n",
      "CPU times: user 5.48 ms, sys: 0 ns, total: 5.48 ms\n",
      "Wall time: 22.5 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "query = {\n",
    "    \"bool\": {\n",
    "        \"must\": {\n",
    "            \"match\": {\"title\": \"Molecular Biology\"}\n",
    "        },\n",
    "        \"filter\": {\n",
    "            \"range\": {\n",
    "                # publish_date\n",
    "                \"date\": {\n",
    "                    \"gte\": \"2022-01-01\",\n",
    "                    \"lte\": \"2023-12-31\"\n",
    "                }\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "}\n",
    "\n",
    "results = search_data(es, index_name, query)\n",
    "\n",
    "print(\"Results: \",len(results), \"\\n\")\n",
    "# Display the results directly\n",
    "for result in results:\n",
    "    # Assuming you want to display each document's key-value pairs\n",
    "    print(result)\n",
    "\n",
    "print(\"-\" * 25)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8afdd444-5566-454f-9ce2-9f156712d3bd",
   "metadata": {},
   "source": [
    "### Aggregation Query\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7be4b39d-4796-4e98-935b-640fb87afb87",
   "metadata": {},
   "source": [
    "Aggregations are pivotal for summarizing data, enabling the analysis of trends across thousands of documents. For example, an aggregation query to count publications by category:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "da813399-4710-40b2-8f17-b4dcdb001e84",
   "metadata": {},
   "outputs": [],
   "source": [
    "query = {\n",
    "    \"size\": 0,  # Important for aggregations, as you might not be interested in the hits but the aggregation results\n",
    "    \"aggs\": {\n",
    "        \"categories\": {  # This name must match what you access in the results\n",
    "            \"terms\": {\n",
    "                \"field\": \"category.keyword\",  # Adjust based on your actual field name\n",
    "                \"size\": 10  # Adjust the size as needed\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "cd0c444b-fd09-4c7d-bcab-cf1b02057e61",
   "metadata": {},
   "outputs": [],
   "source": [
    "def search_data_with_aggregation(es: Elasticsearch, index_name: str, query: Dict[str, Any]) -> List[Dict[str, Any]]:\n",
    "    try:\n",
    "        response = es.search(index=index_name, body=query)\n",
    "        if 'aggregations' in response:\n",
    "            # Safely access the 'categories' aggregation results\n",
    "            categories_agg = response.get('aggregations', {}).get('categories', {}).get('buckets', [])\n",
    "            return categories_agg  # Directly return the categories aggregation results\n",
    "        else:\n",
    "            print(\"No aggregations found in the response.\")\n",
    "            return []\n",
    "    except Exception as e:\n",
    "        print(f\"Search failed: {e}\")\n",
    "        return []"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f2dc134-dc58-4322-9296-29a3bf220ae9",
   "metadata": {},
   "source": [
    "#### Aggregate data, such as counting documents by category."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "905b54ee-bdfa-49c7-8260-2812758f0731",
   "metadata": {},
   "source": [
    "Checks if the response contains an 'aggregations' key and processes the results accordingly. For aggregation queries, it returns the aggregation results directly. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "01fedb95-5ee1-431e-aec0-36c339fbe633",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Category: Neuroscience, Count: 40174\n",
      "Category: Microbiology, Count: 18624\n",
      "Category: Bioinformatics, Count: 18114\n",
      "Category: Cell Biology, Count: 12558\n",
      "Category: Biophysics, Count: 10406\n",
      "Category: Evolutionary Biology, Count: 10384\n",
      "Category: Biochemistry, Count: 8914\n",
      "Category: Immunology, Count: 8834\n",
      "Category: Cancer Biology, Count: 8824\n",
      "Category: Ecology, Count: 8802\n",
      "CPU times: user 2.54 ms, sys: 266 µs, total: 2.8 ms\n",
      "Wall time: 96 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "# Assuming 'es', 'index_name', and 'query' are properly defined\n",
    "results = search_data_with_aggregation(es, index_name, query)\n",
    "\n",
    "# Display the aggregation results\n",
    "if results:\n",
    "    for category in results:\n",
    "        print(f\"Category: {category['key']}, Count: {category['doc_count']}\")\n",
    "else:\n",
    "    print(\"No results found.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a916a837-b9cb-4f29-b87b-1e7a7608aee4",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b4d9fc1b-cfbc-4bc6-811e-316c0759eb4b",
   "metadata": {},
   "source": [
    "### The biorxiv database was downloaded from the medrxivr library"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47bd8fe5-986a-4d1e-a7a3-7154558086dc",
   "metadata": {},
   "source": [
    "The json file used is structure as a list of dictionaries downloaded by MedrdR from a specific date range from 2022 to 2024 and contain 000 papers of data"
   ]
  },
  {
   "cell_type": "raw",
   "id": "da11da8f-11fa-43f2-a447-e9da0b91a9f0",
   "metadata": {},
   "source": [
    "  {\n",
    "    \"doi\": \"10.1101/043794\",\n",
    "    \"title\": \"Modeling methyl-sensitive transcription factor motifs with an expanded epigenetic alphabet\",\n",
    "    \"authors\": \"Viner, C.; Ishak, C. A.; Johnson, J.; Walker, N. J.; Shi, H.; Sjöberg-Herrera, M. K.; Shen, S. Y.; Lardo, S. M.; Adams, D. J.; Ferguson-Smith, A. C.; De Carvalho, D. D.; Hainer, S. J.; Bailey, T. L.; Hoffman, M. M.\",\n",
    "    \"author_corresponding\": \"Michael M. Hoffman\",\n",
    "    \"author_corresponding_institution\": \"Princess Margaret Cancer Centre, Toronto, ON, Canada\",\n",
    "    \"date\": \"2022-07-29\",\n",
    "    \"version\": \"2\",\n",
    "    \"license\": \"cc_by_nc_nd\",\n",
    "    \"category\": \"Bioinformatics\",\n",
    "    \"jatsxml\": \"https://www.biorxiv.org/content/early/2022/07/29/043794.source.xml\",\n",
    "    \"abstract\": \"Transcription factors bind DNA in specific sequence contexts. In addition to distinguishing one nucleobase from another, some transcription factors can distinguish between unmodified and modified bases. Current models of transcription factor binding tend not take DNA modifications into account, while the recent few that do often have limitations. This makes a comprehensive and accurate profiling of transcription factor affinities difficult.\\n\\nHere, we developed methods to identify transcription factor binding sites in modified DNA. Our models expand the standard A/C/G/T DNA alphabet to include cytosine modifications. We developed Cytomod to create modified genomic sequences and enhanced the Multiple EM for Motif Elicitation (MEME) Suite by adding the capacity to handle custom alphabets. We adapted the well-established position weight matrix (PWM) model of transcription factor binding affinity to this expanded DNA alphabet.\\n\\nUsing these methods, we identified modification-sensitive transcription factor binding motifs. We confirmed established binding preferences, such as the preference of ZFP57 and C/EBP{beta} for methylated motifs and the preference of c-Myc for unmethylated E-box motifs. Using known binding preferences to tune model parameters, we discovered novel modified motifs for a wide array of transcription factors. Finally, we validated predicted binding preferences of OCT4 using cleavage under targets and release using nuclease (CUT&RUN) experiments across conventional, methylation-, and hydroxymethylation-enriched sequences. Our approach readily extends to other DNA modifications. As more genome-wide single-base resolution modification data becomes available, we expect that our method will yield insights into altered transcription factor binding affinities across many different modifications.\",\n",
    "    \"published\": \"NA\",\n",
    "    \"node\": 2,\n",
    "    \"link_page\": \"https://www.biorxiv.org/content/10.1101/043794v2?versioned=TRUE\",\n",
    "    \"link_pdf\": \"https://www.biorxiv.org/content/10.1101/043794v2.full.pdf\"\n",
    "  },"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5dbdcbd7-e19d-4c21-9474-1a30e4921d55",
   "metadata": {},
   "source": [
    "### Final Note: The Impact of Advanced Querying on Research\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cff06250-b16f-4f7c-941b-e3d390eaa559",
   "metadata": {},
   "source": [
    "Advanced querying techniques in Elasticsearch empower researchers to navigate and analyze the vast repository of BiorXiv data with unprecedented depth and precision. From basic keyword searches to sophisticated aggregations and scripted calculations, Elasticsearch facilitates a comprehensive understanding of the life sciences landscape. By harnessing these querying capabilities, researchers can accelerate discovery, foster innovation, and contribute to the advancement of science.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "715f4043-82d1-4e8a-98c5-451b478c3e6a",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ccb1cb0-c17d-49aa-89cf-dd365ca3537c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
